---
title: "R Notebook"
output: html_notebook
---

load library
```{r message=FALSE, warning=FALSE}
library(SingleCellExperiment)
library(Seurat)
library(scran)
library(SeuratData)
library(edgeR)
library(limma)
library(clusterProfiler)
library(data.table)
library(SingleCellExperiment)
library(SummarizedExperiment)
library(muscat)
library(pROC)
library(igraph)
library(mclust)
library(cluster)
library(SCINA)
library(caret)
library(MAST)
library(AUCell)
library(HGNChelper)
library(SeuratDisk)
library(monocle3)
library(ggpubr)
library(slingshot)
library(grDevices)
library(RColorBrewer)
library(randomForest)
library(caTools)
library(mltools)
library(stringr)
library(reldist)
library(aricode)
library(TSCAN)
library(reshape2)
library(pheatmap)
library(philentropy)
library(rawr)
library(splatter)
library(combinat)
```

# Correlation between sc and bulk
```{r}
getSCC_sc_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    tmp <- apply(ds[,selectedCells],2,function(x) cor(x,bulk_dat[,status],method="spearman"))
    tmp[is.na(tmp)] <- 0
    return(tmp)
}

getSCC_sc_bulk_diff <- function(ds,col_name,status1,status2){
    selectedCells <- metadat[metadat[,col_name]==status1,]$cell
    tmp <- apply(ds[,selectedCells],2,function(x) cor(x,bulk_dat[,status2],method="spearman"))
    tmp[is.na(tmp)] <- 0
    return(tmp)
}

evalSCC_sc_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    tmp_same <- apply(ds[,selectedCells],2,function(x) cor(x,bulk_dat[,status],method="spearman"))
    tmp_same[is.na(tmp_same)] <- 0
    
    diff_status <- unique(metadat[metadat[,col_name]!=status,col_name])
    tmp_diff <- apply(ds[,selectedCells],2,function(x) max(cor(x,bulk_dat[,diff_status],method="spearman")))
    tmp_diff[is.na(tmp_diff)] <- 0
    
    tmp <- tmp_same-tmp_diff
    return(tmp)
}

getSCC_pb_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    median_PB <- apply(ds[,selectedCells],1,median)
    tmp <- cor.test(median_PB,bulk_dat[,status],method="spearman")
    #tmp_p <- tmp$p.value
    tmp_scc <- tmp$estimate
    tmp_scc[is.na(tmp_scc)] <- 0
    return(tmp_scc)
}

getSCC_pb_bulk_diff <- function(ds,col_name,status1,status2){
    selectedCells <- metadat[metadat[,col_name]==status1,]$cell
    median_PB <- apply(ds[,selectedCells],1,median)
    tmp <- cor.test(median_PB,bulk_dat[,status2],method="spearman")
    #tmp_p <- tmp$p.value
    tmp_scc <- tmp$estimate
    tmp_scc[is.na(tmp_scc)] <- 0
    return(tmp_scc)
}

evalSCC_pb_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    median_PB <- apply(ds[,selectedCells],1,median)
    
    tmp_same <- cor(median_PB,bulk_dat[,status],method="spearman")
    tmp_same[is.na(tmp_same)] <- 0
    
    diff_status <- unique(metadat[metadat[,col_name]!=status,col_name])
    tmp_diff <- max(cor(median_PB,bulk_dat[,diff_status],method="spearman"))
    tmp_diff[is.na(tmp_diff)] <- 0
    
    tmp_scc <- tmp_same-tmp_diff
    return(tmp_scc)
}

getSCC_fc <- function(ds,col_name,status1,status2,LFC_cutoff=0){
    selectedCells1 <- metadat[metadat[,col_name]==status1,]$cell
    median_PB1 <- apply(ds[,selectedCells1],1,median)
    selectedCells2 <- metadat[metadat[,col_name]==status2,]$cell
    median_PB2 <- apply(ds[,selectedCells2],1,median)
    LFC_sc <- median_PB1-median_PB2
    LFC_bulk <- bulk_dat[,status1]-bulk_dat[,status2]
    names(LFC_bulk) <- rownames(bulk_dat)
    UsedGenes <- names(LFC_bulk[abs(LFC_bulk)>=LFC_cutoff])
    tmp <- cor.test(LFC_sc[UsedGenes],LFC_bulk[UsedGenes],method="spearman")
    tmp_scc <- tmp$estimate
    tmp_scc[is.na(tmp_scc)] <- 0
    #tmp_p <- tmp$p.value
    return(tmp_scc)
}

getPsig <- function(x){
    if(x>0.1){
        return(" ")
    }else if(x<=0.0005){
        return("***")
    }else if(x<=0.005){
        return("**")
    }else if(x<=0.05){
        return("*")
    }else if(x<=0.1){
        return(".")
    }
}
```

## GSE75748 cell type
cell level
### process data
```{r message=FALSE, warning=FALSE}
##bulk
bulk_dat <- read.csv("GSE75748_bulk_cell_type_ec.csv",header=TRUE,row.names=1)
bulk_dat <- DGEList(bulk_dat)
keep <- filterByExpr(bulk_dat)
bulk_dat <- bulk_dat[keep,,keep.lib.sizes=FALSE]
bulk_dat <- calcNormFactors(bulk_dat)
bulk_meta <- c("H1","H1","H1","H1","H9","H9","H9","DEC","DEC","EC","EC","EC","HFF","HFF","HFF","NPC","NPC","TB","TB")
mm <- model.matrix(~0 + bulk_meta)
bulk_dat <- voom(bulk_dat,mm,plot=FALSE)$E
bulk_dat <- data.frame(H1=apply(bulk_dat[,1:4],1,mean),H9=apply(bulk_dat[,5:7],1,mean),DEC=apply(bulk_dat[,8:9],1,mean),EC=apply(bulk_dat[,10:12],1,mean),HFF=apply(bulk_dat[,13:15],1,mean),NPC=apply(bulk_dat[,16:17],1,mean),TB=apply(bulk_dat[,18:19],1,mean))

##raw
raw_dat <- fread("rawcount_txt/GSE75748_sc_cell_type_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE75748_sc_cell_type_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE75748_sc_cell_type/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("MAGIC_log/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE75748_sc_cell_type_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE75748_sc_cell_type_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE75748_sc_cell_type_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE75748_sc_cell_type_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE75748_sc_cell_type_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE75748_sc_cell_type_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE75748_sc_cell_type_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE75748_sc_cell_type_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE75748/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE75748_sc_cell_type_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE75748_sc_cell_type_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
GNNimpute_dat <- fread("GNNImpute/GSE75748_sc_cell_type_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE75748_sc_cell_type_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE75748_sc_cell_type_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)


metadat <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat <- data.frame(cell=colnames(raw_dat),celltype=metadat)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat <- list(raw=raw_dat,ADimpute=ADimpute_dat,alra=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,dca=dca_dat,deepimpute=di_dat,GNNimpute=GNNimpute_dat,Iimpute=Iimpute_dat,ks=ks_dat,magic=MAGIC_dat,magiclog=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
bulk_dat <- bulk_dat[rownames(bulk_dat)%in%rownames(dat$dca),]
dat <- lapply(dat,function(x) x[rownames(bulk_dat),] )
```

### relative same-diff SC-Bulk SCC
```{r}
evalSCC_sc_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    tmp_same <- apply(ds[,selectedCells],2,function(x) cor(x,bulk_dat[,status],method="spearman"))
    tmp_same[is.na(tmp_same)] <- 0
    
    diff_status <- unique(metadat[metadat[,col_name]!=status,col_name])
    tmp_diff <- apply(ds[,selectedCells],2,function(x) max(cor(x,bulk_dat[,diff_status],method="spearman")))
    tmp_diff[is.na(tmp_diff)] <- 0
    
    tmp <- tmp_same-tmp_diff
    return(tmp)
}


celltypes <- unique(metadat$celltype)
res_md <- data.frame()
res_pct <- data.frame()
for(x in celltypes){
    tmp <- as.data.frame(lapply(dat,evalSCC_sc_bulk,"celltype",x))
    res_md <- rbind(res_md,apply(tmp,2,median))
    tmp_pct1 <- apply(tmp[,-1],2,function(x) sum(x>=tmp[,1])/nrow(tmp))
    res_pct <- rbind(res_pct,tmp_pct1)
}
res_md <- as.data.frame(t(res_md))
colnames(res_md) <- celltypes
rownames(res_md) <- names(dat)
res_pct <- as.data.frame(t(res_pct))
colnames(res_pct) <- celltypes
rownames(res_pct) <- names(dat)[-1]

#writeout
all_res <- list(cor=res_md,improvePCT=res_pct)
saveRDS(all_res,"PreScreening/Bulk_SC_SpearCor_SameMDiff_GSE75748.rds")
```


### relative same-diff PB-Bulk SCC
```{r}
evalSCC_pb_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    median_PB <- apply(ds[,selectedCells],1,median)
    
    tmp_same <- cor(median_PB,bulk_dat[,status],method="spearman")
    tmp_same[is.na(tmp_same)] <- 0
    
    diff_status <- unique(metadat[metadat[,col_name]!=status,col_name])
    tmp_diff <- max(cor(median_PB,bulk_dat[,diff_status],method="spearman"))
    tmp_diff[is.na(tmp_diff)] <- 0
    
    tmp_scc <- tmp_same-tmp_diff
    return(tmp_scc)
}

celltypes <- unique(metadat$celltype)
res_md <- data.frame()

for(x in celltypes){
    tmp <- as.data.frame(lapply(dat,evalSCC_pb_bulk,"celltype",x))
    res_md <- rbind(res_md,tmp)
}
res_md <- as.data.frame(t(res_md))
colnames(res_md) <- celltypes
rownames(res_md) <- names(dat)

#writeout
all_res <- list(cor=res_md)
saveRDS(all_res,"PreScreening/Bulk_PB_SpearCor_SameMDiff_GSE75748.rds")
```

### PB level fold change between cell type spearman cor
```{r}
celltypes <- unique(metadat$celltype)
comb_celltype <- as.data.frame(combn(celltypes,2))
res_md <- data.frame()
PairGroups <- c()

for(i in 1:ncol(comb_celltype)){
    tmp <- as.data.frame(lapply(dat,getSCC_fc,"celltype",comb_celltype[1,i],comb_celltype[2,i]))
    res_md <- rbind(res_md,tmp)
    PairGroups <- c(PairGroups,paste0(comb_celltype[1,i],"_",comb_celltype[2,i]))
}
res_md <- as.data.frame(t(res_md))
colnames(res_md) <- PairGroups
rownames(res_md) <- names(dat)

#writeout
all_res <- list(cor=res_md)
saveRDS(all_res,"results/PreScreening/Bulk_PB_SpearCor_DiffCellTypeFC_GSE75748.rds")
```
## GSE81861
cell level
### read data
```{r message=FALSE, warning=FALSE}
#cell bench
##bulk
bulk_dat <- readRDS("bulk/bulk_of_GSE81861_with_replicates_TPM.rds")
bulk_dat <- bulk_dat[rowSums(bulk_dat>0)>0,]
bulk_meta <- c("K562","K562","A549","A549","GM12878","GM12878","K562","K562","IMR90","IMR90","H1","H1","GM12878","GM12878") 

bulk_dat <- log2(bulk_dat+1)
bulk_dat <- data.frame(K562=apply(bulk_dat[,c(1,2,7,8)],1,mean),A549=apply(bulk_dat[,3:4],1,mean),GM12878=apply(bulk_dat[,c(5,6,13,14)],1,mean),IMR90=apply(bulk_dat[,9:10],1,mean),H1=apply(bulk_dat[,11:12],1,mean))

##raw
raw_dat <- fread("rawcount_txt/GSE81861_ENCODE_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE81861_ENCODE_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE81861_ENCODE/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("MAGIC_log/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE81861_ENCODE_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE81861_ENCODE_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE81861_ENCODE_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE81861_ENCODE_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE81861_ENCODE_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE81861_ENCODE_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE81861_ENCODE_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE81861_ENCODE_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE81861/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE81861_ENCODE_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE81861_ENCODE_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
# GNNimpute_dat <- fread("GNNImpute/GSE81861_ENCODE_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
# GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE81861_ENCODE_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE81861_ENCODE_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)

metadat <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat <- data.frame(cell=colnames(raw_dat),celltype=metadat)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

#GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat <- list(raw=raw_dat,ADimpute=ADimpute_dat,alra=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,dca=dca_dat,deepimpute=di_dat,Iimpute=Iimpute_dat,ks=ks_dat,magic=MAGIC_dat,magiclog=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
#rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
bulk_dat <- bulk_dat[rownames(bulk_dat)%in%rownames(dat$dca),]
dat <- lapply(dat,function(x) x[rownames(bulk_dat),] )
```

### relative same-diff SC-Bulk SCC
```{r}
evalSCC_sc_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    tmp_same <- apply(ds[,selectedCells],2,function(x) cor(x,bulk_dat[,status],method="spearman"))
    tmp_same[is.na(tmp_same)] <- 0
    
    diff_status <- unique(metadat[metadat[,col_name]!=status,col_name])
    tmp_diff <- apply(ds[,selectedCells],2,function(x) max(cor(x,bulk_dat[,diff_status],method="spearman")))
    tmp_diff[is.na(tmp_diff)] <- 0
    
    tmp <- tmp_same-tmp_diff
    return(tmp)
}


celltypes <- unique(metadat$celltype)
res_md <- data.frame()
res_pct <- data.frame()
for(x in celltypes){
    tmp <- as.data.frame(lapply(dat,evalSCC_sc_bulk,"celltype",x))
    res_md <- rbind(res_md,apply(tmp,2,median))
    tmp_pct1 <- apply(tmp[,-1],2,function(x) sum(x>=tmp[,1])/nrow(tmp))
    res_pct <- rbind(res_pct,tmp_pct1)
}
res_md <- as.data.frame(t(res_md))
colnames(res_md) <- celltypes
rownames(res_md) <- names(dat)
res_pct <- as.data.frame(t(res_pct))
colnames(res_pct) <- celltypes
rownames(res_pct) <- names(dat)[-1]

#writeout
all_res <- list(cor=res_md,improvePCT=res_pct)
saveRDS(all_res,"results/PreScreening/Bulk_SC_SpearCor_SameMDiff_GSE81861.rds")
```


### relative same-diff PB-Bulk SCC
```{r}
evalSCC_pb_bulk <- function(ds,col_name,status){
    selectedCells <- metadat[metadat[,col_name]==status,]$cell
    median_PB <- apply(ds[,selectedCells],1,median)
    
    tmp_same <- cor(median_PB,bulk_dat[,status],method="spearman")
    tmp_same[is.na(tmp_same)] <- 0
    
    diff_status <- unique(metadat[metadat[,col_name]!=status,col_name])
    tmp_diff <- max(cor(median_PB,bulk_dat[,diff_status],method="spearman"))
    tmp_diff[is.na(tmp_diff)] <- 0
    
    tmp_scc <- tmp_same-tmp_diff
    return(tmp_scc)
}

celltypes <- unique(metadat$celltype)
res_md <- data.frame()

for(x in celltypes){
    tmp <- as.data.frame(lapply(dat,evalSCC_pb_bulk,"celltype",x))
    res_md <- rbind(res_md,tmp)
}
res_md <- as.data.frame(t(res_md))
colnames(res_md) <- celltypes
rownames(res_md) <- names(dat)

#writeout
all_res <- list(cor=res_md)
saveRDS(all_res,"results/PreScreening/Bulk_PB_SpearCor_SameMDiff_GSE81861.rds")
```

### PB level fold change between cell type spearman cor
```{r}
celltypes <- unique(metadat$celltype)
comb_celltype <- as.data.frame(combn(celltypes,2))
res_md <- data.frame()
PairGroups <- c()

for(i in 1:ncol(comb_celltype)){
    tmp <- as.data.frame(lapply(dat,getSCC_fc,"celltype",comb_celltype[1,i],comb_celltype[2,i]))
    res_md <- rbind(res_md,tmp)
    PairGroups <- c(PairGroups,paste0(comb_celltype[1,i],"_",comb_celltype[2,i]))
}
res_md <- as.data.frame(t(res_md))
colnames(res_md) <- PairGroups
rownames(res_md) <- names(dat)

#writeout
all_res <- list(cor=res_md)
saveRDS(all_res,"results/PreScreening/Bulk_PB_SpearCor_DiffCellTypeFC_GSE81861.rds")
```

## summary and visualization
### bulk-sc same-diff Correlation
read data
```{r}
d1 <- readRDS("results/PreScreening/Bulk_SC_SpearCor_SameMDiff_GSE75748.rds")
d2 <- readRDS("results/PreScreening/Bulk_SC_SpearCor_SameMDiff_GSE81861.rds")
```

process data
```{r}
tmp1 <- as.data.frame(t(d1$cor))
rownames(tmp1) <- paste0("GSE75748_",rownames(tmp1))

tmp2 <- as.data.frame(t(d2$cor))
rownames(tmp2) <- paste0("GSE81861_",rownames(tmp2))
tmp2$GNNimpute <- NA
tmp2 <- tmp2[,colnames(tmp1)]

tmp <- rbind(tmp1,tmp2)

colnames(tmp)[which(colnames(tmp)=="alra")] <- "ALRA"
colnames(tmp)[which(colnames(tmp)=="dca")] <- "DCA"
colnames(tmp)[which(colnames(tmp)=="Iimpute")] <- "I_Impute"
colnames(tmp)[which(colnames(tmp)=="ks")] <- "kNN_smoothing"
colnames(tmp)[which(colnames(tmp)=="magic")] <- "MAGIC"
colnames(tmp)[which(colnames(tmp)=="magiclog")] <- "MAGIC_log"

md_raw <- median(tmp$raw)
res <- as.data.frame(apply(tmp[,-1],2,function(x) x-tmp[,1] ))
res$celltype <- rownames(res)
res_melt <- data.table::melt(res,id.vars=c("celltype"))

res_melt$Type <- "Deep learning based"
res_melt[res_melt$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
res_melt[res_melt$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
res_melt$Type <- factor(res_melt$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(res_melt)

#rank by cor coefficient diff
mdv <- res_melt %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
res_melt$variable <- factor(res_melt$variable,levels=rank1)
neg_md <- sum(mdv$md<=0)

#performed tests
pv <- apply(tmp[,-1],2,function(x) wilcox.test(tmp[,1],x,alternative="less",paired=TRUE)$p.value )
pv_sig <- sapply(pv,getPsig)
pv_sig <- pv_sig[rank1]
```

box plot
```{r}
tmp_bp <- res_melt[,]

scbulk_smd_scc_bp <- ggboxplot(tmp_bp,x="variable",y="value",fill="Type",ylab="SCC Difference",xlab="Method",title="SC-Bulk Profiling SCC Difference(Relative)")+geom_hline(yintercept=0,linetype="dashed")+geom_vline(xintercept=neg_md+0.5,linetype="dashed")+annotate(geom="text",x=c(1:length(pv_sig)),y=0.15,label=pv_sig,size=7)+annotate(geom="text",x=length(pv_sig)+3,y=0.02,label=paste0("Raw(Median=",round(md_raw,2),")"),size=4.5)+coord_cartesian(xlim=c(1,length(pv_sig)+4.8))+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")
scbulk_smd_scc_bp
```

### bulk-PB same-diff Correlation
read data
```{r}
d1 <- readRDS("results/PreScreening/Bulk_PB_SpearCor_SameMDiff_GSE75748.rds")
d2 <- readRDS("results/PreScreening/Bulk_PB_SpearCor_SameMDiff_GSE81861.rds")
```

process data
```{r}
tmp1 <- as.data.frame(t(d1$cor))
rownames(tmp1) <- paste0("GSE75748_",rownames(tmp1))

tmp2 <- as.data.frame(t(d2$cor))
rownames(tmp2) <- paste0("GSE81861_",rownames(tmp2))
tmp2$GNNimpute <- NA
tmp2 <- tmp2[,colnames(tmp1)]

tmp <- rbind(tmp1,tmp2)

colnames(tmp)[which(colnames(tmp)=="alra")] <- "ALRA"
colnames(tmp)[which(colnames(tmp)=="dca")] <- "DCA"
colnames(tmp)[which(colnames(tmp)=="Iimpute")] <- "I_Impute"
colnames(tmp)[which(colnames(tmp)=="ks")] <- "kNN_smoothing"
colnames(tmp)[which(colnames(tmp)=="magic")] <- "MAGIC"
colnames(tmp)[which(colnames(tmp)=="magiclog")] <- "MAGIC_log"

md_raw <- median(tmp$raw)
res <- as.data.frame(apply(tmp[,-1],2,function(x) x-tmp[,1] ))
res$celltype <- rownames(res)
res_melt <- data.table::melt(res,id.vars=c("celltype"))

res_melt$Type <- "Deep learning based"
res_melt[res_melt$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
res_melt[res_melt$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
res_melt$Type <- factor(res_melt$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(res_melt)

#rank by cor coefficient diff
mdv <- res_melt %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
res_melt$variable <- factor(res_melt$variable,levels=rank1)
neg_md <- sum(mdv$md<=0)

#performed tests
pv <- apply(tmp[,-1],2,function(x) wilcox.test(tmp[,1],x,alternative="less",paired=TRUE)$p.value )
pv_sig <- sapply(pv,getPsig)
pv_sig <- pv_sig[rank1]
```

box plot
```{r}
tmp_bp <- res_melt[,]
#tmp_bp$value <- ifelse(tmp_bp$value<(-0.3),-0.3,tmp_bp$value)
pbbulk_smd_scc_bp <- ggboxplot(tmp_bp,x="variable",y="value",fill="Type",ylab="SCC Difference",xlab="Method",title="Pseudobulk-Bulk Profiling SCC Difference(Relative)")+geom_hline(yintercept=0,linetype="dashed")+geom_vline(xintercept=neg_md+0.5,linetype="dashed")+annotate(geom="text",x=c(1:length(pv_sig)),y=0.16,label=pv_sig,size=7)+annotate(geom="text",x=length(pv_sig)+3,y=0.02,label=paste0("Raw(Median=",round(md_raw,2),")"),size=4.5)+coord_cartesian(xlim=c(1,length(pv_sig)+4.8))+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")
pbbulk_smd_scc_bp
```


### bulk and PB FC different CT Correlation
read data
```{r}
d1 <- readRDS("results/PreScreening/Bulk_PB_SpearCor_DiffCellTypeFC_GSE75748.rds")
d2 <- readRDS("results/PreScreening/Bulk_PB_SpearCor_DiffCellTypeFC_GSE81861.rds")
```

process data
```{r}
tmp1 <- as.data.frame(t(d1$cor))
rownames(tmp1) <- paste0("GSE75748_",rownames(tmp1))

tmp2 <- as.data.frame(t(d2$cor))
rownames(tmp2) <- paste0("GSE81861_",rownames(tmp2))
tmp2$GNNimpute <- NA
tmp2 <- tmp2[,colnames(tmp1)]

tmp <- rbind(tmp1,tmp2)

colnames(tmp)[which(colnames(tmp)=="alra")] <- "ALRA"
colnames(tmp)[which(colnames(tmp)=="dca")] <- "DCA"
colnames(tmp)[which(colnames(tmp)=="Iimpute")] <- "I_Impute"
colnames(tmp)[which(colnames(tmp)=="ks")] <- "kNN_smoothing"
colnames(tmp)[which(colnames(tmp)=="magic")] <- "MAGIC"
colnames(tmp)[which(colnames(tmp)=="magiclog")] <- "MAGIC_log"

md_raw <- median(tmp$raw)
res <- as.data.frame(apply(tmp[,-1],2,function(x) x-tmp[,1] ))
res$celltype <- rownames(res)
res_melt <- data.table::melt(res,id.vars=c("celltype"))

res_melt$Type <- "Deep learning based"
res_melt[res_melt$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
res_melt[res_melt$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
res_melt$Type <- factor(res_melt$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(res_melt)

#rank by cor coefficient diff
mdv <- res_melt %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
res_melt$variable <- factor(res_melt$variable,levels=rank1)
neg_md <- sum(mdv$md<=0)

#performed tests
pv <- apply(tmp[,-1],2,function(x) wilcox.test(tmp[,1],x,alternative="less",paired=TRUE)$p.value )
pv_sig <- sapply(pv,getPsig)
pv_sig <- pv_sig[rank1]
```


Violin plot
```{r}
tmp_bp <- res_melt[,]
#tmp_bp$value <- ifelse(tmp_bp$value<(-0.3),-0.3,tmp_bp$value)
pbbulk_fc_scc_bp <- ggboxplot(tmp_bp,x="variable",y="value",fill="Type",ylab="SCC Difference",xlab="Method",title="Pseudobulk-Bulk Fold Change SCC Difference")+geom_hline(yintercept=0,linetype="dashed")+geom_vline(xintercept=neg_md+0.5,linetype="dashed")+annotate(geom="text",x=c(1:length(pv_sig)),y=0.53,label=pv_sig,size=7)+annotate(geom="text",x=length(pv_sig)+3,y=0.05,label=paste0("Raw(Median=",round(md_raw,2),")"),size=4.5)+coord_cartesian(xlim=c(1,length(pv_sig)+4.8))+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")
pbbulk_fc_scc_bp
```


# DE analysis
functions
```{r}
performDE_bulk <- function(ds,cond,status1,status2){
    #bulk
    mm <- model.matrix(~0 + cond)
    y <- voom(ds,mm,plot=F)
    fit <- lmFit(y, mm)
    #print(head(coef(fit)))
    exp_tmp <- sprintf("cond%s-cond%s",status2,status1)
    myargs <- list(exp_tmp,levels=colnames(coef(fit)))
    contr  <- do.call(makeContrasts,myargs)
    #contr <- makeContrasts(exp_tmp, levels=colnames(coef(fit)))
    tmp <- contrasts.fit(fit,contr)
    tmp <- eBayes(tmp)
    top.table <- topTable(tmp,sort.by="P",n=Inf)
    ## set p=0 to min
    top.table[top.table$P.Value==0,"P.Value"] <- min(top.table[top.table$P.Value!=0,"P.Value"])
    top.table$log10FDR <- log10(top.table$adj.P.Val)*(-1)
    top.table$log10P <- log10(top.table$P.Value)*(-1)
    top.table$signlogFDR <- sign(top.table$logFC)*top.table$log10FDR
    top.table$signlogP <- sign(top.table$logFC)*top.table$log10P
    
    top.table <- top.table[order(top.table[,"signlogP"],abs(top.table[,"logFC"]),decreasing=TRUE),]
    top.table$finalRank <- 1:nrow(top.table)
    return(top.table)
}

#limma trend for pre-normalized data
performDE_bulkTPM <- function(ds,cond,status1,status2){
    #bulk
    mm <- model.matrix(~0 + cond)
    fit <- lmFit(ds,mm)
    #print(head(coef(fit)))
    exp_tmp <- sprintf("cond%s-cond%s",status2,status1)
    myargs <- list(exp_tmp,levels=colnames(coef(fit)))
    contr  <- do.call(makeContrasts,myargs)
    #contr <- makeContrasts(exp_tmp, levels=colnames(coef(fit)))
    tmp <- contrasts.fit(fit,contr)
    tmp <- eBayes(tmp,trend=TRUE)
    top.table <- topTable(tmp,sort.by="P",n=Inf)
    ## set p=0 to min
    top.table[top.table$P.Value==0,"P.Value"] <- min(top.table[top.table$P.Value!=0,"P.Value"])
    top.table$log10FDR <- log10(top.table$adj.P.Val)*(-1)
    top.table$log10P <- log10(top.table$P.Value)*(-1)
    top.table$signlogFDR <- sign(top.table$logFC)*top.table$log10FDR
    top.table$signlogP <- sign(top.table$logFC)*top.table$log10P
    
    top.table <- top.table[order(top.table[,"signlogP"],abs(top.table[,"logFC"]),decreasing=TRUE),]
    top.table$finalRank <- 1:nrow(top.table)
    return(top.table)
}

performDE_sc <- function(ds,meta,label_col,status1,status2,method="rank_sum"){
    
    if(method=="rank_sum"){
        selectedCells <- which(meta[,label_col]%in%c(status1,status2))
        conditions <- factor(meta[selectedCells,label_col],levels=c(status1,status2))
        tmp_ds <- ds[,selectedCells]
        pvalues <- sapply(1:nrow(tmp_ds), function(i){
            dat <- cbind.data.frame(gene=as.numeric(t(tmp_ds[i,])),conditions)
            p <- wilcox.test(gene~conditions, dat)$p.value
            return(p)
        })
        ## set NA to 1
        pvalues <- ifelse(is.na(pvalues),1,pvalues)
        ## set p=0 to min
        pvalues[pvalues==0] <- min(pvalues[pvalues!=0])
        fdr <- p.adjust(pvalues,method="fdr")
        # Calculate the fold-change for each gene
        conditionsLevel <- levels(conditions)
        dataCon1 <- tmp_ds[,c(which(conditions==conditionsLevel[1]))]
        dataCon2 <- tmp_ds[,c(which(conditions==conditionsLevel[2]))]
        foldChanges <- rowMeans(dataCon2)-rowMeans(dataCon1)
        outRst <- data.frame(logFC=foldChanges,P.Value=pvalues,FDR=fdr)
        rownames(outRst) <- rownames(tmp_ds)
        #outRst <- outRst[order(outRst[,"P.Value"],-abs(outRst[,"logFC"])),]
        outRst$log10FDR <- log10(outRst$FDR)*(-1)
        outRst$log10P <- log10(outRst$P.Value)*(-1)
        outRst$signlogFDR <- sign(outRst$logFC)*outRst$log10FDR
        outRst$signlogP <- sign(outRst$logFC)*outRst$log10P
        
        outRst <- outRst[order(outRst[,"signlogP"],abs(outRst[,"logFC"]),decreasing=TRUE),]
        outRst$finalRank <- 1:nrow(outRst)
        outRst <- outRst[rownames(ds),]
        return(outRst)
        
    }else if(method=="MAST"){
        tmp <- FromMatrix(ds,meta,data.frame(features=rownames(ds)),check_sanity=FALSE)
        cdr2 <-colSums(assay(tmp)>0)
        colData(tmp)$cngeneson <- scale(cdr2)

        tmp <- tmp[,which(tmp@colData[,label_col]%in%c(status1,status2))]
        cond <- factor(colData(tmp)[,label_col],levels=c(status1,status2))
        colData(tmp)$cond <- cond
        zlmCond <- zlm(~cond+cngeneson,tmp)
        tmpterm <- paste0("cond",status2)
        summaryCond <- summary(zlmCond,doLRT=tmpterm)
        summaryDt <- as.data.frame(summaryCond$datatable)
        pval <- summaryDt[summaryDt$component=="H"&summaryDt$contrast==tmpterm,c(1,4)]
        lfc <- summaryDt[summaryDt$component=="logFC"&summaryDt$contrast==tmpterm,c(1,7)]
        res <- merge(pval,lfc)
        colnames(res) <- c("Gene","P.Value","logFC")
        rownames(res) <- res$Gene
        #res <- res[order(res[,"P.Value"],-abs(res[,"logFC"])),]
        ## set NA to 1
        res$P.Value <- ifelse(is.na(res$P.Value),1,res$P.Value)
        ## set p=0 to min
        res[res$P.Value==0,"P.Value"] <- min(res[res$P.Value!=0,"P.Value"])
        res$FDR <- p.adjust(res$P.Value,method="fdr")
        res$log10FDR <- log10(res$FDR)*(-1)
        res$log10P <- log10(res$P.Value)*(-1)
        res$signlogFDR <- sign(res$logFC)*res$log10FDR
        res$signlogP <- sign(res$logFC)*res$log10P
        
        res <- res[order(res[,"signlogP"],abs(res[,"logFC"]),decreasing=TRUE),]
        res$finalRank <- 1:nrow(res)
        res <- res[rownames(ds),]
        return(res)
    }
}

getSCC <- function(sc,bulk){
    res <- cor.test(sc,bulk,method="pearson")
    #return(list(estimate=res$estimate,p=res$p.value))
    return(res$estimate)
}

getSCC2 <- function(sc,bulk){
    #selectedG <- which(!is.nan(sc)) 
    #sc <- sc[selectedG]
    #bulk <- bulk[selectedG]
    res <- tryCatch(
        {
            res <- cor.test(sc,bulk,method="spearman")
            #return(list(estimate=res$estimate,p=res$p.value))
            return(res$estimate)
        },
        error=function(cond){
            return(NA)
        }
    )
    return(res)
}

getOverlappedDEGs <- function(scd,top_n,allgenes,bulk_degs){
    names(scd) <- allgenes
    top_degs <- names(scd[order(scd)])[1:top_n]
    return(sum(top_degs%in%bulk_degs)/top_n)
}

getOverlappedDEGs2 <- function(scd,top_n,allgenes,bulk_degs){
    names(scd) <- allgenes
    top_degs <- names(scd[order(scd)])[1:top_n]
    return(sum(top_degs%in%bulk_degs)/length(bulk_degs))
}

getJaccard <- function(scd,top_n,allgenes,bulk_degs){
    names(scd) <- allgenes
    top_degs <- names(scd[order(scd)])[1:top_n]
    interS <- length(intersect(top_degs,bulk_degs))
    unionS <- length(top_degs)+length(bulk_degs)-interS
    return(interS/unionS)
}

getAllSCC <- function(bulkds,scds){
    # bulk_res1 <- performDE_bulk(bulkds,bulkconds,status1,status2)
    # sc_res1 <- as.data.frame(lapply(scds,performDE_sc,scmeta,label_col,status1,status2))
    sc_res1 <- scds[rownames(bulkds),]
    bulk_res1 <- bulkds[,]
    #correlation between rank
    tmp1 <- sc_res1[,grep("finalRank",colnames(sc_res1))]
    tmp1 <- apply(tmp1,2,getSCC,bulk_res1[,"finalRank"])
    tmp1 <- as.data.frame(tmp1)
    #0.3
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.3,])
    tmp2 <- sc_res1[SelectedG,grep("finalRank",colnames(sc_res1))]
    tmp2 <- apply(tmp2,2,getSCC,bulk_res1[SelectedG,"finalRank"])
    tmp2 <- as.data.frame(tmp2)
    #0.15
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.15,])
    tmp3 <- sc_res1[SelectedG,grep("finalRank",colnames(sc_res1))]
    tmp3 <- apply(tmp3,2,getSCC,bulk_res1[SelectedG,"finalRank"])
    tmp3 <- as.data.frame(tmp3)
    #0.05
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.05,])
    tmp4 <- sc_res1[SelectedG,grep("finalRank",colnames(sc_res1))]
    tmp4 <- apply(tmp4,2,getSCC,bulk_res1[SelectedG,"finalRank"])
    tmp4 <- as.data.frame(tmp4)
    #0.005
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.005,])
    tmp5 <- sc_res1[SelectedG,grep("finalRank",colnames(sc_res1))]
    tmp5 <- apply(tmp5,2,getSCC,bulk_res1[SelectedG,"finalRank"])
    tmp5 <- as.data.frame(tmp5)

    tmp_p <- cbind(tmp1,tmp2,tmp3,tmp4,tmp5)
    colnames(tmp_p) <- c("all","0.3","0.15","0.05","0.005")
    
    #correlation between logFC
    tmp1 <- sc_res1[,grep("logFC",colnames(sc_res1))]
    tmp1 <- apply(tmp1,2,getSCC2,bulk_res1[,"logFC"])
    tmp1 <- as.data.frame(tmp1)
    #0.3
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.3,])
    tmp2 <- sc_res1[SelectedG,grep("logFC",colnames(sc_res1))]
    tmp2 <- apply(tmp2,2,getSCC2,bulk_res1[SelectedG,"logFC"])
    tmp2 <- as.data.frame(tmp2)
    #0.15
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.15,])
    tmp3 <- sc_res1[SelectedG,grep("logFC",colnames(sc_res1))]
    tmp3 <- apply(tmp3,2,getSCC2,bulk_res1[SelectedG,"logFC"])
    tmp3 <- as.data.frame(tmp3)
    #0.05
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.05,])
    tmp4 <- sc_res1[SelectedG,grep("logFC",colnames(sc_res1))]
    tmp4 <- apply(tmp4,2,getSCC2,bulk_res1[SelectedG,"logFC"])
    tmp4 <- as.data.frame(tmp4)
    #0.05
    SelectedG <- rownames(bulk_res1[bulk_res1$P.Value<=0.005,])
    tmp5 <- sc_res1[SelectedG,grep("logFC",colnames(sc_res1))]
    tmp5 <- apply(tmp5,2,getSCC2,bulk_res1[SelectedG,"logFC"])
    tmp5 <- as.data.frame(tmp5)
    
    tmp_fc <- cbind(tmp1,tmp2,tmp3,tmp4,tmp5)
    colnames(tmp_fc) <- c("all","0.3","0.15","0.05","0.005")
    
    #overlaps between top100,200,500,1000DEGs
    #100
    bulk_DEGs1 <- rownames(bulk_res1[order(bulk_res1$P.Value),])[1:100]
    sc_DEGs <- sc_res1[,grep("P.Value",colnames(sc_res1))]
    sc_DEGs1 <- apply(sc_DEGs,2,getOverlappedDEGs,100,rownames(sc_res1),bulk_DEGs1)
    sc_DEGs1 <- as.data.frame(sc_DEGs1)
    sc_jaccard1 <- apply(sc_DEGs,2,getJaccard,100,rownames(sc_res1),bulk_DEGs1)
    sc_jaccard1 <- as.data.frame(sc_jaccard1)
    
    #200
    bulk_DEGs2 <- rownames(bulk_res1[order(bulk_res1$P.Value),])[1:200]
    sc_DEGs <- sc_res1[,grep("P.Value",colnames(sc_res1))]
    sc_DEGs2 <- apply(sc_DEGs,2,getOverlappedDEGs,200,rownames(sc_res1),bulk_DEGs2)
    sc_DEGs2 <- as.data.frame(sc_DEGs2)
    sc_jaccard2 <- apply(sc_DEGs,2,getJaccard,200,rownames(sc_res1),bulk_DEGs2)
    sc_jaccard2 <- as.data.frame(sc_jaccard2)
    
    #500
    bulk_DEGs3 <- rownames(bulk_res1[order(bulk_res1$P.Value),])[1:500]
    sc_DEGs <- sc_res1[,grep("P.Value",colnames(sc_res1))]
    sc_DEGs3 <- apply(sc_DEGs,2,getOverlappedDEGs,500,rownames(sc_res1),bulk_DEGs3)
    sc_DEGs3 <- as.data.frame(sc_DEGs3)
    sc_jaccard3 <- apply(sc_DEGs,2,getJaccard,500,rownames(sc_res1),bulk_DEGs3)
    sc_jaccard3 <- as.data.frame(sc_jaccard3)
    
    #1000(no need)
    bulk_DEGs4 <- rownames(bulk_res1[order(bulk_res1$P.Value),])[1:1000]
    sc_DEGs <- sc_res1[,grep("P.Value",colnames(sc_res1))]
    sc_DEGs4 <- apply(sc_DEGs,2,getOverlappedDEGs,1000,rownames(sc_res1),bulk_DEGs4)
    sc_DEGs4 <- as.data.frame(sc_DEGs4)
    sc_jaccard4 <- apply(sc_DEGs,2,getJaccard,1000,rownames(sc_res1),bulk_DEGs4)
    sc_jaccard4 <- as.data.frame(sc_jaccard4)
    
    tmp_overlap <- cbind(sc_DEGs1,sc_DEGs2,sc_DEGs3,sc_DEGs4)
    colnames(tmp_overlap) <- c("top100","top200","top500","top1000")
    
    tmp_jaccard <- cbind(sc_jaccard1,sc_jaccard2,sc_jaccard3,sc_jaccard4)
    colnames(tmp_jaccard) <- c("top100","top200","top500","top1000")
    
    return(list(p=tmp_p,fc=tmp_fc,overlap=tmp_overlap,jaccard=tmp_jaccard))
}

getAllOverlap <- function(bulkds,scds,nbulkDEGs=100,nscDEGs=100*c(1:5)){
    
    sc_DEGs <- scds[,grep("P.Value",colnames(scds))]
    bulk_DEGs <- rownames(bulkds[order(bulkds$P.Value),])

    sc_DEGs1 <- lapply(nscDEGs,function(x) as.data.frame(apply(sc_DEGs,2,getOverlappedDEGs2,x,rownames(sc_DEGs),bulk_DEGs[1:nbulkDEGs])))
    sc_DEGs1 <- as.data.frame(sc_DEGs1)
    colnames(sc_DEGs1) <- paste0("scTOP",nscDEGs)
    
    sc_jaccard1 <- lapply(nscDEGs,function(x) as.data.frame(apply(sc_DEGs,2,getJaccard,x,rownames(sc_DEGs),bulk_DEGs[1:nbulkDEGs])))
    sc_jaccard1 <- as.data.frame(sc_jaccard1)
    colnames(sc_jaccard1) <- paste0("scTOP",nscDEGs)
    
    return(list(overlap=sc_DEGs1,jaccard=sc_jaccard1))
}

performPBDE_sc_RST <- function(ds,meta,label_col,sample_col,cellt_col,status1,status2){
    tmp <- SingleCellExperiment(ds,colData=meta,mainExpName="log2norm")
    tmp <- prepSCE(tmp,kid=cellt_col,gid=label_col,sid=sample_col,drop=FALSE)
    assayNames(tmp) <- "log2norm"
    tmp <- aggregateData(tmp,assay="log2norm",fun="mean",by=c("cluster_id","sample_id"))
    conds <- tmp$group_id

    #Rank sum
    tmpdat <- tmp@assays@data[[1]]
    pvalues <- sapply(1:nrow(tmpdat), function(i){
        dat <- cbind.data.frame(gene=as.numeric(t(tmpdat[i,])),conds)
        p <- wilcox.test(gene~conds, dat)$p.value
        return(p)
    })
    ## set NA to 1
    pvalues <- ifelse(is.na(pvalues),1,pvalues)
    ## set p=0 to min
    pvalues[pvalues==0] <- min(pvalues[pvalues!=0])
    fdr <- p.adjust(pvalues, method="fdr")
    # Calculate the fold-change for each gene
    #conditionsLevel <- levels(conds)
    dataCon1 <- tmpdat[,c(which(conds==status1))]
    dataCon2 <- tmpdat[,c(which(conds==status2))]
    foldChanges <- rowMeans(dataCon2)-rowMeans(dataCon1)
    outRst <- data.frame(logFC=foldChanges,P.Value=pvalues,FDR=fdr)
    rownames(outRst) <- rownames(tmpdat)
    outRst$log10FDR <- log10(outRst$FDR)*(-1)
    outRst$log10P <- log10(outRst$P.Value)*(-1)
    outRst$signlogFDR <- sign(outRst$logFC)*outRst$log10FDR
    outRst$signlogP <- sign(outRst$logFC)*outRst$log10P
    
    outRst <- outRst[order(outRst[,"signlogP"],abs(outRst[,"logFC"]),decreasing=TRUE),]
    outRst$finalRank <- 1:nrow(outRst)
    outRst <- outRst[rownames(ds),]
    return(outRst)
}

#limma trend for pre-normalized data
performPBDE_sc_limma <- function(ds,meta,label_col,sample_col,cellt_col,status1,status2){
    tmp <- SingleCellExperiment(ds,colData=meta,mainExpName="log2norm")
    tmp <- prepSCE(tmp,kid=cellt_col,gid=label_col,sid=sample_col,drop=FALSE)
    assayNames(tmp) <- "log2norm"
    tmp <- aggregateData(tmp,assay="log2norm",fun="mean",by=c("cluster_id","sample_id"))
    conds <- tmp$group_id

    #limma
    mm <- model.matrix(~0 + conds)
    fit <- lmFit(tmp@assays@data[[1]],mm)
    exp_tmp <- sprintf("conds%s-conds%s",status2,status1)
    myargs <- list(exp_tmp,levels=colnames(coef(fit)))
    contr  <- do.call(makeContrasts,myargs)
    tmp <- contrasts.fit(fit,contr)
    tmp <- eBayes(tmp,trend=TRUE)
    top.table <- topTable(tmp,sort.by="P",n=Inf)
    ## set NA to 1
    top.table$P.Value <- ifelse(is.na(top.table$P.Value),1,top.table$P.Value)
    ## set p=0 to min
    top.table$P.Value[top.table$P.Value==0] <- min(top.table$P.Value[top.table$P.Value!=0])
    top.table$log10FDR <- log10(top.table$adj.P.Val)*(-1)
    top.table$log10P <- log10(top.table$P.Value)*(-1)
    top.table$signlogFDR <- sign(top.table$logFC)*top.table$log10FDR
    top.table$signlogP <- sign(top.table$logFC)*top.table$log10P
    top.table <- top.table[order(top.table[,"signlogP"],abs(top.table[,"logFC"]),decreasing=TRUE),]
    top.table$finalRank <- 1:nrow(top.table)
    top.table <- top.table[rownames(ds),]
    return(top.table)
}

getPsig <- function(x){
    if(x>0.1){
        return(" ")
    }else if(x<=0.0005){
        return("***")
    }else if(x<=0.005){
        return("**")
    }else if(x<=0.05){
        return("*")
    }else if(x<=0.1){
        return(".")
    }
}
```

## GSE75748 cell type
cell level
read data
```{r message=FALSE, warning=FALSE}
#cell bench
##bulk
bulk_dat <- read.csv("bulk/GSE75748_bulk_cell_type_ec.csv",header=TRUE,row.names=1)
bulk_dat <- DGEList(bulk_dat)
keep <- filterByExpr(bulk_dat)
bulk_dat <- bulk_dat[keep,,keep.lib.sizes=FALSE]
bulk_dat <- calcNormFactors(bulk_dat)
bulk_meta <- c("H1","H1","H1","H1","H9","H9","H9","DEC","DEC","EC","EC","EC","HFF","HFF","HFF","NPC","NPC","TB","TB") 
#mm <- model.matrix(~0 + bulk_meta)
#bulk_dat <- voom(bulk_dat,mm,plot=FALSE)$E

##raw
raw_dat <- fread("rawcount_txt/GSE75748_sc_cell_type_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE75748_sc_cell_type_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE75748_sc_cell_type/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("MAGIC_log/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE75748_sc_cell_type_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE75748_sc_cell_type_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE75748_sc_cell_type_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE75748_sc_cell_type_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE75748_sc_cell_type_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE75748_sc_cell_type_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE75748_sc_cell_type_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE75748_sc_cell_type_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE75748/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE75748_sc_cell_type_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE75748_sc_cell_type_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
GNNimpute_dat <- fread("GNNImpute/GSE75748_sc_cell_type_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE75748_sc_cell_type_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE75748_sc_cell_type_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)

metadat <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat <- data.frame(cell=colnames(raw_dat),celltype=metadat)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat <- list(raw=raw_dat,ADimpute=ADimpute_dat,alra=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,dca=dca_dat,deepimpute=di_dat,GNNimpute=GNNimpute_dat,Iimpute=Iimpute_dat,ks=ks_dat,magic=MAGIC_dat,magiclog=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
colnames(dat$alra) <- rownames(metadat)
bulk_dat <- bulk_dat[rownames(bulk_dat)%in%rownames(dat$dca),]
dat <- lapply(dat,function(x) x[rownames(bulk_dat),] )
```

### single cell level
#### result rank sum
```{r message=FALSE, warning=FALSE}
status <- unique(metadat$celltype)
comb_status <- as.data.frame(combn(status,2))
res <- list()

#ncol(comb_status)
for(i in 1:ncol(comb_status)){
    bulk_res1 <- performDE_bulk(bulk_dat,bulk_meta,comb_status[1,i],comb_status[2,i])
    sc_res1 <- as.data.frame(lapply(dat,performDE_sc,metadat,"celltype",comb_status[1,i],comb_status[2,i]))
    sc_res1 <- sc_res1[rownames(bulk_res1),]
    PairGroups <- paste0(comb_status[1,i],"_",comb_status[2,i])
    res[[paste0("bulk_",PairGroups)]] <- bulk_res1
    res[[paste0("sc_",PairGroups)]] <- sc_res1
}

#writeout
saveRDS(res,"results/PreScreening/Bulk_limma_SC_RankSum_DE_GSE75748.rds")
```

## GSE81861
cell level
read data
```{r message=FALSE, warning=FALSE}
##bulk
bulk_dat <- readRDS("bulk/bulk_of_GSE81861_with_replicates_TPM.rds")
bulk_dat <- bulk_dat[rowSums(bulk_dat>0)>0,]
filterGene <- which(apply(bulk_dat,1,max)<5)
bulk_dat <- bulk_dat[-filterGene,]
bulk_meta <- c("K562","K562","A549","A549","GM12878","GM12878","K562","K562","IMR90","IMR90","H1","H1","GM12878","GM12878") 
bulk_dat <- log2(bulk_dat+1)

##raw
raw_dat <- fread("rawcount_txt/GSE81861_ENCODE_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE81861_ENCODE_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE81861_ENCODE/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("imputed/MAGIC_log/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE81861_ENCODE_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE81861_ENCODE_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE81861_ENCODE_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE81861_ENCODE_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE81861_ENCODE_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE81861_ENCODE_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE81861_ENCODE_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE81861_ENCODE_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE81861/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE81861_ENCODE_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE81861_ENCODE_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
# GNNimpute_dat <- fread("/restricted/projectnb/casa/jh50/jinghan/others/SCI/imputed/GNNImpute/GSE81861_ENCODE_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
# GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE81861_ENCODE_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE81861_ENCODE_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)

metadat <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat <- data.frame(cell=colnames(raw_dat),celltype=metadat)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

#GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat <- list(raw=raw_dat,ADimpute=ADimpute_dat,alra=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,dca=dca_dat,deepimpute=di_dat,Iimpute=Iimpute_dat,ks=ks_dat,magic=MAGIC_dat,magiclog=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
#rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
bulk_dat <- bulk_dat[rownames(bulk_dat)%in%rownames(dat$raw),]
dat <- lapply(dat,function(x) x[rownames(bulk_dat),] )
```

### single cell level
#### result rank sum
```{r message=FALSE, warning=FALSE}
status <- unique(metadat$celltype)
comb_status <- as.data.frame(combn(status,2))
res <- list()

#ncol(comb_status)
for(i in 1:ncol(comb_status)){
    bulk_res1 <- performDE_bulkTPM(bulk_dat,bulk_meta,comb_status[1,i],comb_status[2,i])
    sc_res1 <- as.data.frame(lapply(dat,performDE_sc,metadat,"celltype",comb_status[1,i],comb_status[2,i]))
    sc_res1 <- sc_res1[rownames(bulk_res1),]
    PairGroups <- paste0(comb_status[1,i],"_",comb_status[2,i])
    res[[paste0("bulk_",PairGroups)]] <- bulk_res1
    res[[paste0("sc_",PairGroups)]] <- sc_res1
}

#writeout
saveRDS(res,"results/PreScreening/Bulk_limma_SC_RankSum_DE_GSE81861.rds")
```

## summary and visualization
### P/FC sc-bulk correlations
read data
```{r message=FALSE, warning=FALSE}
tmp <- list.files("results/PreScreening/")
tmp <- tmp[grep("Bulk_limma_SC_RankSum_DE",tmp)]

#GSE757548
res1 <- list()
tmp1 <- tmp[grep("DE_GSE75748",tmp)]
for(x in tmp1){
    tmppath <- paste0("results/PreScreening/",x)
    dat <- readRDS(tmppath)
    res1 <- c(res1,dat)
}

Compare_names <- unique(sapply(names(res1),function(x) strsplit(x,"bulk|sc")[[1]][2] ))
#get results
p1_all <- data.frame()
p1_005 <- data.frame()
fc1_all <- data.frame()
fc1_005 <- data.frame()
for(x in Compare_names){
    bulkres <- paste0("bulk",x)
    scres <- paste0("sc",x)
    tmpres <- getAllSCC(res1[[bulkres]],res1[[scres]])
    method_names <- rownames(tmpres$p)
    p1_all <- rbind(p1_all,tmpres$p$all)
    p1_005 <- rbind(p1_005,tmpres$p$`0.05`)
    fc1_all <- rbind(fc1_all,tmpres$fc$all)
    fc1_005 <- rbind(fc1_005,tmpres$fc$`0.05`)
}

method_names <- sapply(method_names,function(x) strsplit(x,"\\.")[[1]][1] )
Compare_names <- paste0("GSE75748",Compare_names)
rownames(p1_all) <- Compare_names
colnames(p1_all) <- method_names
rownames(p1_005) <- Compare_names
colnames(p1_005) <- method_names
rownames(fc1_all) <- Compare_names
colnames(fc1_all) <- method_names
rownames(fc1_005) <- Compare_names
colnames(fc1_005) <- method_names

#GSE81861
res2 <- list()
tmp2 <- tmp[grep("DE_GSE81861",tmp)]
for(x in tmp2){
    tmppath <- paste0("results/PreScreening/",x)
    dat <- readRDS(tmppath)
    res2 <- c(res2,dat)
}

Compare_names <- unique(sapply(names(res2),function(x) strsplit(x,"bulk|sc")[[1]][2] ))
#get results
p2_all <- data.frame()
p2_005 <- data.frame()
fc2_all <- data.frame()
fc2_005 <- data.frame()
for(x in Compare_names){
    bulkres <- paste0("bulk",x)
    scres <- paste0("sc",x)
    tmpres <- getAllSCC(res2[[bulkres]],res2[[scres]])
    method_names <- rownames(tmpres$p)
    p2_all <- rbind(p2_all,tmpres$p$all)
    p2_005 <- rbind(p2_005,tmpres$p$`0.05`)
    fc2_all <- rbind(fc2_all,tmpres$fc$all)
    fc2_005 <- rbind(fc2_005,tmpres$fc$`0.05`)
}

method_names <- sapply(method_names,function(x) strsplit(x,"\\.")[[1]][1] )
Compare_names <- paste0("GSE81861",Compare_names)
rownames(p2_all) <- Compare_names
colnames(p2_all) <- method_names
rownames(p2_005) <- Compare_names
colnames(p2_005) <- method_names
rownames(fc2_all) <- Compare_names
colnames(fc2_all) <- method_names
rownames(fc2_005) <- Compare_names
colnames(fc2_005) <- method_names

#merge
p2_all$GNNimpute <- NA
p2_005$GNNimpute <- NA
fc2_all$GNNimpute <- NA
fc2_005$GNNimpute <- NA
p2_all <- p2_all[,colnames(p1_all)]
p2_005 <- p2_005[,colnames(p1_005)]
fc2_all <- fc2_all[,colnames(fc1_all)]
fc2_005 <- fc2_005[,colnames(fc1_005)]

p_all <- rbind(p1_all,p2_all)
p_005 <- rbind(p1_005,p2_005)
fc_all <- rbind(fc1_all,fc2_all)
fc_005 <- rbind(fc1_005,fc2_005)
```

process data
```{r}
colnames(p_all)[which(colnames(p_all)=="alra")] <- "ALRA"
colnames(p_all)[which(colnames(p_all)=="dca")] <- "DCA"
colnames(p_all)[which(colnames(p_all)=="Iimpute")] <- "I_Impute"
colnames(p_all)[which(colnames(p_all)=="ks")] <- "kNN_smoothing"
colnames(p_all)[which(colnames(p_all)=="magic")] <- "MAGIC"
colnames(p_all)[which(colnames(p_all)=="magiclog")] <- "MAGIC_log"

md_raw_p <- median(p_all$raw)
p_all_res <- as.data.frame(apply(p_all[,-1],2,function(x) x-p_all[,1] ))
p_all_res$group <- rownames(p_all_res)

colnames(fc_all)[which(colnames(fc_all)=="alra")] <- "ALRA"
colnames(fc_all)[which(colnames(fc_all)=="dca")] <- "DCA"
colnames(fc_all)[which(colnames(fc_all)=="Iimpute")] <- "I_Impute"
colnames(fc_all)[which(colnames(fc_all)=="ks")] <- "kNN_smoothing"
colnames(fc_all)[which(colnames(fc_all)=="magic")] <- "MAGIC"
colnames(fc_all)[which(colnames(fc_all)=="magiclog")] <- "MAGIC_log"

md_raw_fc <- median(fc_all$raw)
fc_all_res <- as.data.frame(apply(fc_all[,-1],2,function(x) x-fc_all[,1] ))
fc_all_res$group <- rownames(fc_all_res)

p_all_res_melt <- data.table::melt(p_all_res,id.vars=c("group"))

p_all_res_melt$Type <- "Deep learning based"
p_all_res_melt[p_all_res_melt$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
p_all_res_melt[p_all_res_melt$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
p_all_res_melt$Type <- factor(p_all_res_melt$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(p_all_res_melt)
#p_005_res_melt <- data.table::melt(p_005_res,id.vars=c("group"))
fc_all_res_melt <- data.table::melt(fc_all_res,id.vars=c("group"))

fc_all_res_melt$Type <- "Deep learning based"
fc_all_res_melt[fc_all_res_melt$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
fc_all_res_melt[fc_all_res_melt$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
fc_all_res_melt$Type <- factor(fc_all_res_melt$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

```

Violin plot
p_all_res
```{r}
mdv <- p_all_res_melt %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
p_all_res_melt$variable <- factor(p_all_res_melt$variable,levels=rank1)
neg_md <- sum(mdv$md<=0)

#performed tests
pv <- apply(p_all[,-1],2,function(x) wilcox.test(p_all[,1],x,alternative="less",paired=TRUE)$p.value )
pv_sig <- sapply(pv,getPsig)
pv_sig <- pv_sig[rank1]

tmp_bp <- p_all_res_melt[,]
tmp_bp$value <- ifelse(tmp_bp$value<(-0.1),-0.1,tmp_bp$value)

de_p_all_scc_bp <- ggboxplot(tmp_bp,x="variable",y="value",fill="Type",ylab="SCC Difference",xlab="Method",title="Bulk-SC Gene Sign(P) Rank SCC Difference(Rank Sum)")+geom_hline(yintercept=0,linetype="dashed")+geom_vline(xintercept=neg_md+0.5,linetype="dashed")+annotate(geom="text",x=c(1:length(pv_sig)),y=0.15,label=pv_sig,size=7)+annotate(geom="text",x=length(pv_sig)+3,y=0.02,label=paste0("Raw(Median=",round(md_raw_p,2),")"),size=4.5)+coord_cartesian(xlim=c(1,length(pv_sig)+4.8))+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")

de_p_all_scc_bp
```

fc_all_res
```{r}
mdv <- fc_all_res_melt %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
fc_all_res_melt$variable <- factor(fc_all_res_melt$variable,levels=rank1)
neg_md <- sum(mdv$md<=0)

#performed tests
pv <- apply(fc_all[,-1],2,function(x) wilcox.test(fc_all[,1],x,alternative="less",paired=TRUE)$p.value )
pv_sig <- sapply(pv,getPsig)
pv_sig <- pv_sig[rank1]

tmp_bp <- fc_all_res_melt[,]
tmp_bp$value <- ifelse(tmp_bp$value<(-0.1),-0.1,tmp_bp$value)
de_fc_all_scc_bp <- ggboxplot(tmp_bp,x="variable",y="value",fill="Type",ylab="SCC Difference",xlab="Method",title="Bulk-SC Gene logFC SCC Difference")+geom_hline(yintercept=0,linetype="dashed")+geom_vline(xintercept=neg_md+0.5,linetype="dashed")+annotate(geom="text",x=c(1:length(pv_sig)),y=0.28,label=pv_sig,size=7)+annotate(geom="text",x=length(pv_sig)+3,y=0.02,label=paste0("Raw(Median=",round(md_raw_fc,2),")"),size=4.5)+coord_cartesian(xlim=c(1,length(pv_sig)+4.8))+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")
de_fc_all_scc_bp
```

### improve significance
read data
```{r message=FALSE, warning=FALSE}
tmp <- list.files("results/PreScreening/")
tmp <- tmp[grep("Bulk_limma_SC_RankSum_DE",tmp)]

#GSE757548
res1 <- list()
tmp1 <- tmp[grep("DE_GSE75748",tmp)]
for(x in tmp1){
    tmppath <- paste0("results/PreScreening/",x)
    dat <- readRDS(tmppath)
    res1 <- c(res1,dat)
}

Compare_names <- unique(sapply(names(res1),function(x) strsplit(x,"bulk|sc")[[1]][2] ))
#get results
sig_all <- data.frame()
for(x in Compare_names){
    bulkres <- paste0("bulk",x)
    scres <- paste0("sc",x)
    selectedGenes <- rownames(res1[[bulkres]][order(res1[[bulkres]]$P.Value),])[1:500]
    topgenesP_sc <- res1[[scres]][selectedGenes,grep("P.Value",colnames(res1[[scres]]))]
    method_names <- colnames(topgenesP_sc)
    topgenesFC_sc <- res1[[scres]][selectedGenes,grep("logFC",colnames(res1[[scres]]))]
    improve_sig <- apply(topgenesP_sc[,-1],2,function(x) x<=topgenesP_sc[,1])
    same_dir <- apply(topgenesFC_sc[,-1],2,function(x) x*res1[[bulkres]][selectedGenes,]$logFC>0)
    improve_sig <- improve_sig + same_dir
    improve_sig <- apply(improve_sig,2,function(x) sum(x==2)/length(x))
    sig_all <- rbind(sig_all,improve_sig)
}
method_names <- sapply(method_names,function(x) strsplit(x,"\\.")[[1]][1] )
Compare_names <- paste0("GSE75748",Compare_names)
rownames(sig_all) <- Compare_names
colnames(sig_all) <- method_names[-1]

#GSE81861
res2 <- list()
tmp2 <- tmp[grep("DE_GSE81861",tmp)]
for(x in tmp2){
    tmppath <- paste0("results/PreScreening/",x)
    dat <- readRDS(tmppath)
    res2 <- c(res2,dat)
}

Compare_names <- unique(sapply(names(res2),function(x) strsplit(x,"bulk|sc")[[1]][2] ))
#get results
sig_all2 <- data.frame()
for(x in Compare_names){
    bulkres <- paste0("bulk",x)
    scres <- paste0("sc",x)
    selectedGenes <- rownames(res2[[bulkres]][order(res2[[bulkres]]$P.Value),])[1:500]
    topgenesP_sc <- res2[[scres]][selectedGenes,grep("P.Value",colnames(res2[[scres]]))]
    method_names <- colnames(topgenesP_sc)
    topgenesFC_sc <- res2[[scres]][selectedGenes,grep("logFC",colnames(res2[[scres]]))]
    improve_sig <- apply(topgenesP_sc[,-1],2,function(x) x<=topgenesP_sc[,1])
    same_dir <- apply(topgenesFC_sc[,-1],2,function(x) x*res2[[bulkres]][selectedGenes,]$logFC>0)
    improve_sig <- improve_sig + same_dir
    improve_sig <- apply(improve_sig,2,function(x) sum(x==2)/length(x))
    sig_all2 <- rbind(sig_all2,improve_sig)
}
method_names <- sapply(method_names,function(x) strsplit(x,"\\.")[[1]][1] )
Compare_names <- paste0("GSE81861",Compare_names)
rownames(sig_all2) <- Compare_names
colnames(sig_all2) <- method_names[-1]

#merge
sig_all2$GNNimpute <- NA
sig_all2 <- sig_all2[,colnames(sig_all)]

sig_all <- rbind(sig_all,sig_all2)
```

process data
```{r}
colnames(sig_all)[which(colnames(sig_all)=="alra")] <- "ALRA"
colnames(sig_all)[which(colnames(sig_all)=="dca")] <- "DCA"
colnames(sig_all)[which(colnames(sig_all)=="Iimpute")] <- "I_Impute"
colnames(sig_all)[which(colnames(sig_all)=="ks")] <- "kNN_smoothing"
colnames(sig_all)[which(colnames(sig_all)=="magic")] <- "MAGIC"
colnames(sig_all)[which(colnames(sig_all)=="magiclog")] <- "MAGIC_log"

sig_all$group <- rownames(sig_all)
sig_all <- data.table::melt(sig_all,id.vars=c("group"))

sig_all$Type <- "Deep learning based"
sig_all[sig_all$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
sig_all[sig_all$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
sig_all$Type <- factor(sig_all$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(sig_all)
```

Violin plot
sig_all
```{r}
#rank
mdv <- sig_all %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
sig_all$Method <- as.character(sig_all$variable)
sig_all$variable <- factor(sig_all$variable,levels=rank1)
neg_md <- sum(mdv$md<=0.5)

de_top_improve_bp_rst <- ggboxplot(sig_all,x="variable",y="value",fill="Type")+geom_hline(yintercept=0.5,linetype="dashed")+geom_vline(xintercept=neg_md+0.5,linetype="dashed")+xlab("Method")+ylab("Improved,%")+ggtitle("% of top500 DEGs with improved significance(Rank Sum)")+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")
de_top_improve_bp_rst
```

### FP DEGs
read data
```{r message=FALSE, warning=FALSE}
tmp <- list.files("results/PreScreening/")
tmp <- tmp[grep("Bulk_limma_SC_RankSum_DE",tmp)]

#GSE757548
res1 <- list()
tmp1 <- tmp[grep("DE_GSE75748",tmp)]
for(x in tmp1){
    tmppath <- paste0("results/PreScreening/",x)
    dat <- readRDS(tmppath)
    res1 <- c(res1,dat)
}

Compare_names <- unique(sapply(names(res1),function(x) strsplit(x,"bulk|sc")[[1]][2] ))
#get results
fp_rate1 <- data.frame()
for(x in Compare_names){
    bulkres <- paste0("bulk",x)
    scres <- paste0("sc",x)
    selectedGenes <- tail(rownames(res1[[bulkres]][order(res1[[bulkres]]$P.Value),]),round(nrow(res1[[bulkres]])/3))
    false_pos <- res1[[scres]][,grep("P.Value",colnames(res1[[scres]]))]
    method_names <- colnames(false_pos)
    false_pos <- apply(false_pos,2,getOverlappedDEGs,500,rownames(res1[[scres]]),selectedGenes)
    #false_pos <- as.data.frame(false_pos)
    fp_rate1 <- rbind(fp_rate1,false_pos)
}
method_names <- sapply(method_names,function(x) strsplit(x,"\\.")[[1]][1] )
Compare_names <- paste0("GSE75748",Compare_names)
rownames(fp_rate1) <- Compare_names
colnames(fp_rate1) <- method_names

#GSE81861
res2 <- list()
tmp2 <- tmp[grep("DE_GSE81861",tmp)]
for(x in tmp2){
    tmppath <- paste0("results/PreScreening/",x)
    dat <- readRDS(tmppath)
    res2 <- c(res2,dat)
}

Compare_names <- unique(sapply(names(res2),function(x) strsplit(x,"bulk|sc")[[1]][2] ))
#get results
fp_rate2 <- data.frame()
for(x in Compare_names){
    bulkres <- paste0("bulk",x)
    scres <- paste0("sc",x)
    selectedGenes <- tail(rownames(res2[[bulkres]][order(res2[[bulkres]]$P.Value),]),round(nrow(res2[[bulkres]])/3))
    false_pos <- res2[[scres]][,grep("P.Value",colnames(res2[[scres]]))]
    method_names <- colnames(false_pos)
    false_pos <- apply(false_pos,2,getOverlappedDEGs,500,rownames(res2[[scres]]),selectedGenes)
    #false_pos <- as.data.frame(false_pos)
    fp_rate2 <- rbind(fp_rate2,false_pos)
}
method_names <- sapply(method_names,function(x) strsplit(x,"\\.")[[1]][1] )
Compare_names <- paste0("GSE81861",Compare_names)
rownames(fp_rate2) <- Compare_names
colnames(fp_rate2) <- method_names

#merge
fp_rate2$GNNimpute <- NA
fp_rate2 <- fp_rate2[,colnames(fp_rate1)]

fp_rate <- rbind(fp_rate1,fp_rate2)

colnames(fp_rate)[which(colnames(fp_rate)=="alra")] <- "ALRA"
colnames(fp_rate)[which(colnames(fp_rate)=="dca")] <- "DCA"
colnames(fp_rate)[which(colnames(fp_rate)=="Iimpute")] <- "I_Impute"
colnames(fp_rate)[which(colnames(fp_rate)=="ks")] <- "kNN_smoothing"
colnames(fp_rate)[which(colnames(fp_rate)=="magic")] <- "MAGIC"
colnames(fp_rate)[which(colnames(fp_rate)=="magiclog")] <- "MAGIC_log"
```

process data
```{r}
md_raw <- median(fp_rate$raw)
fp_rate_diff <- as.data.frame(apply(fp_rate[,-1],2,function(x) x-fp_rate[,1] ))
fp_rate_diff$group <- rownames(fp_rate_diff)
fp_rate_diff_melt <- data.table::melt(fp_rate_diff,id.vars=c("group"))

fp_rate_diff_melt$Type <- "Deep learning based"
fp_rate_diff_melt[fp_rate_diff_melt$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
fp_rate_diff_melt[fp_rate_diff_melt$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
fp_rate_diff_melt$Type <- factor(fp_rate_diff_melt$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(fp_rate_diff_melt)
```

Violin plot
fp_rate
```{r}
mdv <- fp_rate_diff_melt %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
fp_rate_diff_melt$Method <- as.character(fp_rate_diff_melt$variable)
fp_rate_diff_melt$variable <- factor(fp_rate_diff_melt$variable,levels=rank1)
neg_md <- sum(mdv$md<=0)

#performed tests
pv <- apply(fp_rate[,-1],2,function(x) wilcox.test(fp_rate[,1],x,alternative="two.sided",paired=TRUE)$p.value )
pv_sig <- sapply(pv,getPsig)
pv_sig <- pv_sig[as.character(rank1)]

tmp_bp <- fp_rate_diff_melt[,]
tmp_bp$value <- ifelse(tmp_bp$value>(0.2),0.2,tmp_bp$value)
de_fp_bp <- ggboxplot(tmp_bp,x="variable",y="value",fill="Type")+geom_hline(yintercept=0,linetype="dashed")+xlab("Method")+ylab("FP rate Difference")+ggtitle("False Positive Rate Difference(Rank Sum)")+annotate(geom="text",x=c(1:length(pv_sig)),y=0.22,label=pv_sig,size=7)+annotate(geom="text",x=length(pv_sig)+3.0,y=0.03,label=paste0("Raw(Median=",round(md_raw,2),")"),size=4.5)+coord_cartesian(xlim=c(1,length(pv_sig)+5.2))+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")
de_fp_bp
```


### AUC predictive performance
read GSE75748 data
```{r message=FALSE, warning=FALSE}
#cell bench
##bulk
bulk_dat <- read.csv("bulk/GSE75748_bulk_cell_type_ec.csv",header=TRUE,row.names=1)
bulk_dat <- DGEList(bulk_dat)
keep <- filterByExpr(bulk_dat)
bulk_dat <- bulk_dat[keep,,keep.lib.sizes=FALSE]
bulk_dat <- calcNormFactors(bulk_dat)
bulk_meta <- c("H1","H1","H1","H1","H9","H9","H9","DEC","DEC","EC","EC","EC","HFF","HFF","HFF","NPC","NPC","TB","TB") 
#mm <- model.matrix(~0 + bulk_meta)
#bulk_dat <- voom(bulk_dat,mm,plot=FALSE)$E

##raw
raw_dat <- fread("rawcount_txt/GSE75748_sc_cell_type_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE75748_sc_cell_type_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE75748_sc_cell_type/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("MAGIC_log/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE75748_sc_cell_type_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE75748_sc_cell_type_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE75748_sc_cell_type_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE75748_sc_cell_type_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE75748_sc_cell_type_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE75748_sc_cell_type_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE75748_sc_cell_type_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE75748_sc_cell_type_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE75748/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE75748_sc_cell_type_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE75748_sc_cell_type_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
GNNimpute_dat <- fread("GNNImpute/GSE75748_sc_cell_type_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE75748_sc_cell_type_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE75748_sc_cell_type_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)

metadat <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat <- data.frame(cell=colnames(raw_dat),celltype=metadat)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat <- list(raw=raw_dat,ADimpute=ADimpute_dat,ALRA=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,DCA=dca_dat,deepimpute=di_dat,GNNimpute=GNNimpute_dat,I_Impute=Iimpute_dat,kNN_smoothing=ks_dat,MAGIC=MAGIC_dat,MAGIC_log=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
colnames(dat$ALRA) <- rownames(metadat)
bulk_dat <- bulk_dat[rownames(bulk_dat)%in%rownames(dat$DCA),]
dat <- lapply(dat,function(x) x[rownames(bulk_dat),] )
```

read GSE81861
```{r message=FALSE, warning=FALSE}
##bulk
bulk_dat2 <- readRDS("bulk/bulk_of_GSE81861_with_replicates_TPM.rds")
bulk_dat2 <- bulk_dat2[rowSums(bulk_dat2>0)>0,]
filterGene <- which(apply(bulk_dat2,1,max)<5)
bulk_dat2 <- bulk_dat2[-filterGene,]
bulk_meta2 <- c("K562","K562","A549","A549","GM12878","GM12878","K562","K562","IMR90","IMR90","H1","H1","GM12878","GM12878") 
bulk_dat2 <- log2(bulk_dat2+1)

##raw
raw_dat <- fread("rawcount_txt/GSE81861_ENCODE_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE81861_ENCODE_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE81861_ENCODE/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("MAGIC_log/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE81861_ENCODE_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE81861_ENCODE_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE81861_ENCODE_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE81861_ENCODE_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE81861_ENCODE_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE81861_ENCODE_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE81861_ENCODE_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE81861_ENCODE_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE81861/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE81861_ENCODE_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE81861_ENCODE_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
# GNNimpute_dat <- fread("/restricted/projectnb/casa/jh50/jinghan/others/SCI/imputed/GNNImpute/GSE81861_ENCODE_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
# GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE81861_ENCODE_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE81861_ENCODE_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)

metadat2 <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat2 <- data.frame(cell=colnames(raw_dat),celltype=metadat2)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat2,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat2,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat2,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat2,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

#GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat2 <- list(raw=raw_dat,ADimpute=ADimpute_dat,ALRA=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,DCA=dca_dat,deepimpute=di_dat,I_Impute=Iimpute_dat,kNN_smoothing=ks_dat,MAGIC=MAGIC_dat,MAGIC_log=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
#rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
bulk_dat2 <- bulk_dat2[rownames(bulk_dat2)%in%rownames(dat2$raw),]
dat2 <- lapply(dat2,function(x) x[rownames(bulk_dat2),] )
```

```{r message=FALSE, warning=FALSE}
getAUC <- function(x,outcome){
    tmp <- auc(outcome,x)
    tmp <- attr(tmp,"roc")$auc
    return(tmp)
}

#GSE75748
celltypes <- unique(metadat$celltype)
comb_celltype <- as.data.frame(combn(celltypes,2))
auc_res1 <- data.frame()
for(cb in 1:ncol(comb_celltype)){
    selectedGenes <- rownames(performDE_bulk(bulk_dat,bulk_meta,comb_celltype[1,cb],comb_celltype[2,cb]))[1:10]
    selectedCells <- metadat[metadat$celltype%in%c(comb_celltype[1,cb],comb_celltype[2,cb]),"cell"]
    dat_AUC <- lapply(dat,function(x) t(x[selectedGenes,selectedCells]))
    pred_perform <- as.data.frame(lapply(dat_AUC,function(x) 
        apply(x,2,getAUC,metadat[selectedCells,"celltype"])))
    rownames(pred_perform) <- 
        paste0(rownames(pred_perform),"_",comb_celltype[1,cb],"_",comb_celltype[2,cb],"_GSE75748")
    auc_res1 <- rbind(auc_res1,pred_perform)
}

#GSE81861
celltypes <- unique(metadat2$celltype)
comb_celltype <- as.data.frame(combn(celltypes,2))
auc_res2 <- data.frame()
for(cb in 1:ncol(comb_celltype)){
    selectedGenes <- rownames(performDE_bulkTPM(bulk_dat2,bulk_meta2,comb_celltype[1,cb],comb_celltype[2,cb]))[1:10]
    selectedCells <- metadat2[metadat2$celltype%in%c(comb_celltype[1,cb],comb_celltype[2,cb]),"cell"]
    dat_AUC <- lapply(dat2,function(x) t(x[selectedGenes,selectedCells]))
    pred_perform <- as.data.frame(lapply(dat_AUC,function(x) 
        apply(x,2,getAUC,metadat2[selectedCells,"celltype"])))
    rownames(pred_perform) <- 
        paste0(rownames(pred_perform),"_",comb_celltype[1,cb],"_",comb_celltype[2,cb],"_GSE81861")
    auc_res2 <- rbind(auc_res2,pred_perform)
}

#merge
auc_res2$GNNimpute <- NA
auc_res2 <- auc_res2[,colnames(auc_res1)]

auc_res <- rbind(auc_res1,auc_res2)
```

process data
```{r}
md_raw <- median(auc_res$raw)
auc_res_diff <- as.data.frame(apply(auc_res[,-1],2,function(x) x-auc_res[,1] ))
auc_res_diff$group <- rownames(auc_res_diff)
auc_res_diff <- data.table::melt(auc_res_diff,id.vars=c("group"))

auc_res_diff$Type <- "Deep learning based"
auc_res_diff[auc_res_diff$variable%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
auc_res_diff[auc_res_diff$variable%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
auc_res_diff$Type <- factor(auc_res_diff$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(auc_res_diff)
```

Violin plot(all)
auc_res_diff
```{r}
mdv <- auc_res_diff %>% dplyr::group_by(variable) %>% dplyr::summarise(md=median(value,na.rm=TRUE))
rank1 <- mdv[order(mdv$md),]$variable
auc_res_diff$variable <- factor(auc_res_diff$variable,levels=rank1)
neg_md <- sum(mdv$md<=0)

#performed tests
pv <- apply(auc_res[,-1],2,function(x) wilcox.test(auc_res[,1],x,alternative="less",paired=TRUE)$p.value )
pv_sig <- sapply(pv,getPsig)
pv_sig <- pv_sig[rank1]

tmp_bp <- auc_res_diff[,]
#tmp_bp$value <- ifelse(tmp_bp$value<(-0.1),-0.1,tmp_bp$value)
tmp_title <- paste0("Biomarker(All;n=",nrow(auc_res),") Predictive Performance-AUC Difference")
pred_all_auc_bp <- ggboxplot(tmp_bp,x="variable",y="value",fill="Type")+geom_hline(yintercept=0,linetype="dashed")+geom_vline(xintercept=neg_md+0.5,linetype="dashed")+xlab("Method")+ylab("AUC Difference")+ggtitle(tmp_title)+annotate(geom="text",x=length(pv_sig)+3,y=0.06,label=paste0("Raw(Median=",round(md_raw,2),")"),size=4.5)+coord_cartesian(xlim=c(1,length(pv_sig)+4.8))+annotate(geom="text",x=c(1:length(pv_sig)),y=0.7,label=pv_sig,size=7)+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))+scale_fill_brewer(palette="Set2")
pred_all_auc_bp
```

### classification
read GSE75748 data
```{r message=FALSE, warning=FALSE}
#cell bench
##bulk
bulk_dat <- read.csv("bulk/GSE75748_bulk_cell_type_ec.csv",header=TRUE,row.names=1)
bulk_dat <- DGEList(bulk_dat)
keep <- filterByExpr(bulk_dat)
bulk_dat <- bulk_dat[keep,,keep.lib.sizes=FALSE]
bulk_dat <- calcNormFactors(bulk_dat)
bulk_meta <- c("H1","H1","H1","H1","H9","H9","H9","DEC","DEC","EC","EC","EC","HFF","HFF","HFF","NPC","NPC","TB","TB") 
#mm <- model.matrix(~0 + bulk_meta)
#bulk_dat <- voom(bulk_dat,mm,plot=FALSE)$E

##raw
raw_dat <- fread("rawcount_txt/GSE75748_sc_cell_type_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE75748_sc_cell_type_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE75748_sc_cell_type/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("MAGIC_log/GSE75748_sc_cell_type_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE75748_sc_cell_type_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE75748_sc_cell_type_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE75748_sc_cell_type_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE75748_sc_cell_type_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE75748_sc_cell_type_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE75748_sc_cell_type_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE75748_sc_cell_type_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE75748_sc_cell_type_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE75748/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE75748_sc_cell_type_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE75748_sc_cell_type_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE75748_sc_cell_type_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
GNNimpute_dat <- fread("imputed/GNNImpute/GSE75748_sc_cell_type_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE75748_sc_cell_type_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE75748_sc_cell_type_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)

metadat <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat <- data.frame(cell=colnames(raw_dat),celltype=metadat)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat <- list(raw=raw_dat,ADimpute=ADimpute_dat,ALRA=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,DCA=dca_dat,deepimpute=di_dat,GNNimpute=GNNimpute_dat,I_Impute=Iimpute_dat,kNN_smoothing=ks_dat,MAGIC=MAGIC_dat,MAGIC_log=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
colnames(dat$ALRA) <- rownames(metadat)
bulk_dat <- bulk_dat[rownames(bulk_dat)%in%rownames(dat$DCA),]
dat <- lapply(dat,function(x) x[rownames(bulk_dat),] )
```

read GSE81861
```{r message=FALSE, warning=FALSE}
##bulk
bulk_dat2 <- readRDS("bulk/bulk_of_GSE81861_with_replicates_TPM.rds")
bulk_dat2 <- bulk_dat2[rowSums(bulk_dat2>0)>0,]
filterGene <- which(apply(bulk_dat2,1,max)<5)
bulk_dat2 <- bulk_dat2[-filterGene,]
bulk_meta2 <- c("K562","K562","A549","A549","GM12878","GM12878","K562","K562","IMR90","IMR90","H1","H1","GM12878","GM12878") 
bulk_dat2 <- log2(bulk_dat2+1)

##raw
raw_dat <- fread("rawcount_txt/GSE81861_ENCODE_genebycell.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
raw_dat <- as.matrix(raw_dat,rownames=1)
##imputed
MAGIC_dat <- fread("MAGIC/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIC_dat <- as.matrix(MAGIC_dat,rownames=1)
alra_dat <- fread("alra/GSE81861_ENCODE_genebycell_alra.txt",sep="\t",stringsAsFactors=FALSE)
alra_dat <- as.matrix(alra_dat,rownames=1)
dca_dat <- fread("dca/GSE81861_ENCODE/mean.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
dca_dat <- as.matrix(dca_dat,rownames=1)
MAGIClog_dat <- fread("MAGIC_log/GSE81861_ENCODE_genebycell.tsv",sep="\t",header=TRUE,stringsAsFactors=FALSE)
MAGIClog_dat <- as.matrix(MAGIClog_dat,rownames=1)
afMF_dat <- fread("afMF/imputed_data/GSE81861_ENCODE_sigma3_0_convergence_True.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
afMF_dat <- as.matrix(afMF_dat,rownames=1)
ADimpute_dat <- fread("ADImpute/GSE81861_ENCODE_genebycell_ADI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ADimpute_dat <- as.matrix(ADimpute_dat,rownames=1)
AutoClass_dat <- fread("AutoClass/GSE81861_ENCODE_genebycell_AC.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
AutoClass_dat <- as.matrix(AutoClass_dat,rownames=1)
ccImpute_dat <- fread("ccImpute/GSE81861_ENCODE_genebycell_cc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ccImpute_dat <- as.matrix(ccImpute_dat,rownames=1)
di_dat <- fread("deepimpute/GSE81861_ENCODE_genebycell_di.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
di_dat <- as.matrix(di_dat,rownames=1)
ks_dat <- fread("knn_smoothing/GSE81861_ENCODE_genebycell_ks.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
ks_dat <- as.matrix(ks_dat,rownames=1)
scIGAN_dat <- fread("scIGANs/scIGANs_GSE81861_ENCODE_genebycell.txt_noLabel.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scIGAN_dat <- as.matrix(scIGAN_dat,rownames=1)
Iimpute_dat <- fread("I_Impute/GSE81861_ENCODE_genebycell_CImp.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Iimpute_dat <- as.matrix(Iimpute_dat,rownames=1)
scGNN2_dat <- fread("scGNN2/GSE81861/imputed.csv",sep=",",header=TRUE,stringsAsFactors=FALSE)
scGNN2_dat <- as.matrix(scGNN2_dat,rownames=1)
scGNN2_dat <- t(scGNN2_dat)
scRMDraw_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDraw.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDraw_dat <- as.matrix(scRMDraw_dat,rownames=1)
scRMDnorm_dat <- fread("scRMD/GSE81861_ENCODE_genebycell_scRMDlognorm.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scRMDnorm_dat <- as.matrix(scRMDnorm_dat,rownames=1)
SMURF_dat <- fread("SMURF/GSE81861_ENCODE_genebycell_SMURF.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
SMURF_dat <- as.matrix(SMURF_dat,rownames=1)
Bf_dat <- fread("Bfimpute/GSE81861_ENCODE_genebycell_Bf_specc.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
Bf_dat <- as.matrix(Bf_dat,rownames=1)
# GNNimpute_dat <- fread("/restricted/projectnb/casa/jh50/jinghan/others/SCI/imputed/GNNImpute/GSE81861_ENCODE_genebycell_GNN.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
# GNNimpute_dat <- as.matrix(GNNimpute_dat,rownames=1)
RESCUE_dat <- fread("RESCUE/GSE81861_ENCODE_genebycell_RESCUE.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
RESCUE_dat <- as.matrix(RESCUE_dat,rownames=1)
scVI_dat <- fread("scVI/GSE81861_ENCODE_genebycell_scVI.txt",sep="\t",header=TRUE,stringsAsFactors=FALSE)
scVI_dat <- as.matrix(scVI_dat,rownames=1)

metadat2 <- sapply(colnames(raw_dat),function(x) strsplit(x,"\\_")[[1]][1])
metadat2 <- data.frame(cell=colnames(raw_dat),celltype=metadat2)
```

normalized/log2 data
```{r}
raw_dat <- CreateSeuratObject(counts=raw_dat,meta.data=metadat2,min.cells=0,min.features=0)
raw_dat <- NormalizeData(raw_dat)
raw_dat <- as.matrix(raw_dat@assays$RNA@data)

dca_dat <- log2(dca_dat)

ks_dat <- CreateSeuratObject(counts=ks_dat,meta.data=metadat2,min.cells=0,min.features=0)
ks_dat <- NormalizeData(ks_dat)
ks_dat <- as.matrix(ks_dat@assays$RNA@data)

scIGAN_dat <- log2(scIGAN_dat)

di_dat <- CreateSeuratObject(counts=di_dat,meta.data=metadat2,min.cells=0,min.features=0)
di_dat <- NormalizeData(di_dat)
di_dat <- as.matrix(di_dat@assays$RNA@data)

Iimpute_dat <- log2(Iimpute_dat)

scRMDraw_dat <- CreateSeuratObject(counts=scRMDraw_dat,meta.data=metadat2,min.cells=0,min.features=0)
scRMDraw_dat <- NormalizeData(scRMDraw_dat)
scRMDraw_dat <- as.matrix(scRMDraw_dat@assays$RNA@data)

SMURF_dat <- log2(SMURF_dat+min(as.numeric(SMURF_dat[SMURF_dat!=0])))

#GNNimpute_dat <- log2(GNNimpute_dat+min(as.numeric(GNNimpute_dat[GNNimpute_dat!=0])))

scVI_dat <- log2(scVI_dat)
```

combine normalized data
```{r}
dat2 <- list(raw=raw_dat,ADimpute=ADimpute_dat,ALRA=alra_dat,AutoClass=AutoClass_dat,Bfimpute=Bf_dat,ccImpute=ccImpute_dat,DCA=dca_dat,deepimpute=di_dat,I_Impute=Iimpute_dat,kNN_smoothing=ks_dat,MAGIC=MAGIC_dat,MAGIC_log=MAGIClog_dat,RESCUE=RESCUE_dat,scGNN2=scGNN2_dat,scIGANs=scIGAN_dat,scRMDraw=scRMDraw_dat,scRMDnorm=scRMDnorm_dat,scVI=scVI_dat,SMURF=SMURF_dat,afMF=afMF_dat)
rm(raw_dat)
rm(alra_dat)
rm(ADimpute_dat)
rm(AutoClass_dat)
rm(Bf_dat)
rm(ccImpute_dat)
rm(di_dat)
#rm(GNNimpute_dat)
rm(Iimpute_dat)
rm(ks_dat)
rm(dca_dat)
rm(MAGIC_dat)
rm(MAGIClog_dat)
rm(RESCUE_dat)
rm(scGNN2_dat)
rm(scIGAN_dat)
rm(scRMDraw_dat)
rm(scRMDnorm_dat)
rm(scVI_dat)
rm(SMURF_dat)
rm(afMF_dat)

#remove inconsistant genes
bulk_dat2 <- bulk_dat2[rownames(bulk_dat2)%in%rownames(dat2$raw),]
dat2 <- lapply(dat2,function(x) x[rownames(bulk_dat2),] )
```

function
```{r}
evalAcc <- function(rf,label){
    acc <- sum(rf$test$predicted==label)/length(label)
    return(acc)
}

evalmedianF1 <- function(rf,label){
    tmp <- confusionMatrix(factor(label,levels=levels(rf$test$predicted)),rf$test$predicted,mode="everything")
    tmp <- median(tmp$byClass[,"F1"],na.rm=TRUE)
    return(tmp)
}

evalProb <- function(rf){
    maxprob <- apply(rf$test$votes,1,max)
    return(maxprob)
}

evalF1 <- function(rf,label){
    tmp <- confusionMatrix(factor(label,levels=levels(rf$test$predicted)),rf$test$predicted,mode="everything")
    tmp <- tmp$byClass[,"F1"]
    return(tmp)
}

evalF1TwoClass <- function(rf,label){
    tmp <- confusionMatrix(factor(label,levels=levels(rf$test$predicted)),rf$test$predicted,mode="everything")
    tmp <- tmp$byClass["F1"]
    return(tmp)
}

trainModel <- function(x,selectedG,meta,labelcol,splitsp){
    tmp <- x
    tmp <- tmp[,selectedG]
    tmp$label <- meta[,labelcol]
    colnames(tmp) <- str_replace(colnames(tmp),"-","_")
    trainset <- subset(tmp,splitsp==TRUE)
    testset <- subset(tmp,splitsp==FALSE)
    rf <- randomForest(x=trainset[,1:(ncol(trainset)-1)],y=as.factor(trainset$label),xtest=testset[,1:(ncol(testset)-1)],ytest=as.factor(testset$label))
    return(list(rf=rf,label=testset$label))
}
```


```{r message=FALSE, warning=FALSE}
#GSE75748
bulk_meta_c <- as.factor(bulk_meta)
bulk_meta_c <- one_hot(as.data.table(bulk_meta_c))
bulk_meta_c <- as.data.frame( cbind(bulk_meta,bulk_meta_c) )
bulk_meta_c <- as.data.frame(apply(bulk_meta_c,2,function(x) as.character(x)))

selectedGenes <- c()
for(x in unique(bulk_meta)){
    y <- paste0("bulk_meta_c_",x)
    tmp <- rownames(performDE_bulk(bulk_dat,bulk_meta_c[,y],"0","1"))[1:10]
    selectedGenes <- c(selectedGenes,tmp)
}

selectedGenes <- unique(selectedGenes)

dat_c <- lapply(dat,function(x) as.data.frame(t(x)))
set.seed(1)
splitsample <- sample.split(metadat$celltype,SplitRatio=0.7)
table(metadat[splitsample,]$celltype)
table(metadat[!splitsample,]$celltype)

class_res <- lapply(dat_c,trainModel,selectedGenes,metadat,"celltype",splitsample)

##accracy
res_acc1 <- as.data.frame(lapply(class_res,function(x) evalAcc(x$rf,x$label) ))
res_acc1 <- as.data.frame(t(res_acc1))

#cell type probablibility
res_prob1 <- as.data.frame(lapply(class_res,function(x) evalProb(x$rf) ))
res_truect1 <- as.data.frame(lapply(class_res,function(x) x$rf$test$predicted==x$label))
res_truect1 <- as.data.frame(apply(res_truect1[,-1],2,function(x) x*res_truect1[,1])) ##both raw and imputed are true label
rownames(res_truect1) <- rownames(res_prob1)

#GSE81861
bulk_meta_c <- as.factor(bulk_meta2)
bulk_meta_c <- one_hot(as.data.table(bulk_meta_c))
bulk_meta_c <- as.data.frame( cbind(bulk_meta2,bulk_meta_c) )
bulk_meta_c <- as.data.frame(apply(bulk_meta_c,2,function(x) as.character(x)))

selectedGenes <- c()
for(x in unique(bulk_meta2)){
    y <- paste0("bulk_meta_c_",x)
    tmp <- rownames(performDE_bulkTPM(bulk_dat2,bulk_meta_c[,y],"0","1"))[1:10]
    selectedGenes <- c(selectedGenes,tmp)
}

selectedGenes <- unique(selectedGenes)

dat_c <- lapply(dat2,function(x) as.data.frame(t(x)))
set.seed(1)
splitsample <- sample.split(metadat2$celltype,SplitRatio=0.7)
table(metadat2[splitsample,]$celltype)
table(metadat2[!splitsample,]$celltype)

class_res <- lapply(dat_c,trainModel,selectedGenes,metadat2,"celltype",splitsample)

##accracy
res_acc2 <- as.data.frame(lapply(class_res,function(x) evalAcc(x$rf,x$label) ))
res_acc2 <- as.data.frame(t(res_acc2))

#cell type probablibility
res_prob2 <- as.data.frame(lapply(class_res,function(x) evalProb(x$rf) ))
res_truect2 <- as.data.frame(lapply(class_res,function(x) x$rf$test$predicted==x$label))
res_truect2 <- as.data.frame(apply(res_truect2[,-1],2,function(x) x*res_truect2[,1])) ##both raw and imputed are true label
rownames(res_truect2) <- rownames(res_prob2)

#merge
#auc_res2$GNNimpute <- NA
res_acc2 <- res_acc2[rownames(res_acc1),]
res_acc <- cbind(res_acc1,res_acc2)

res_prob2$GNNimpute <- NA
res_prob2 <- res_prob2[,colnames(res_prob1)]
res_prob <- rbind(res_prob1,res_prob2)
res_truect2$GNNimpute <- NA
res_truect2 <- res_truect2[,colnames(res_truect1)]
res_truect <- rbind(res_truect1,res_truect2)

res_hprob <- apply(res_prob[,-1],2,function(x) x>res_prob[,1])
res_hprob <- res_hprob*res_truect
res_hprob <- sapply(c(1:(length(dat)-1)),function(i) 
    sum(res_hprob[,i],na.rm=TRUE)/sum(res_truect[,i],na.rm=TRUE))
#res_hprob1 <- apply(res_hprob1,2,function(x) sum(x)/nrow(res_hprob1))

res_eprob <- apply(res_prob[,-1],2,function(x) x==res_prob[,1])
res_eprob <- res_eprob*res_truect
res_eprob <- sapply(c(1:(length(dat)-1)),function(i) 
    sum(res_eprob[,i],na.rm=TRUE)/sum(res_truect[,i],na.rm=TRUE))
#lower ct max prob
res_lprob <- 1-res_hprob-res_eprob

res_allprob <- data.frame(higher=res_hprob,equal=res_eprob,lower=res_lprob)
rownames(res_allprob) <- colnames(res_prob[,-1])

```

process data
```{r}
res_acc_mean <- as.data.frame(apply(res_acc,1,mean,na.rm=TRUE))
res_acc_mean$method <- rownames(res_acc_mean)
colnames(res_acc_mean)[1] <- "mean_acc"

res_acc_mean$Type <- "Deep learning based"
res_acc_mean[res_acc_mean$method%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
res_acc_mean[res_acc_mean$method%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
res_acc_mean[res_acc_mean$method%in%c("raw"),]$Type <- "Raw"
res_acc_mean$Type <- factor(res_acc_mean$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based","Raw"))

head(res_acc_mean)

res_allprob$method <- rownames(res_allprob)
res_allprob_melt <- data.table::melt(res_allprob,id.vars=c("method"))

res_allprob_melt$Type <- "Deep learning based"
res_allprob_melt[res_allprob_melt$method%in%c("ADimpute","ccImpute","I_Impute","kNN_smoothing","MAGIC","MAGIC_log","RESCUE"),]$Type <- "Model or smoothing based"
res_allprob_melt[res_allprob_melt$method%in%c("ALRA","Bfimpute","afMF","scRMDraw","scRMDnorm"),]$Type <- "Matrix based"
res_allprob_melt$Type <- factor(res_allprob_melt$Type,levels=c("Matrix based","Model or smoothing based","Deep learning based"))

head(res_allprob_melt)
```

res_acc_mean
```{r}
rank1 <- res_acc_mean[order(res_acc_mean$mean_acc),]$method
#res_acc_mean$Method <- factor(res_acc_mean$method,levels=c("raw",res_acc_mean$method[-1][order(res_acc_mean$method[-1])]))
res_acc_mean$method <- factor(res_acc_mean$method,levels=rank1)

class_acc_bp <- ggbarplot(res_acc_mean,x="method",y="mean_acc",fill="Type",xlab="Method",ylab="Accuracy")+coord_cartesian(ylim=c(min(res_acc_mean$mean_acc),1))+ggtitle("Random Forest Cell Type Classification Accuracy")+geom_hline(yintercept=res_acc_mean[res_acc_mean$method=="raw",]$mean_acc,linetype="dashed")+scale_fill_brewer(palette="Set2")+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))
class_acc_bp
```

stacked bar plot for prob
res_allprob_melt
```{r}
rank1 <- res_allprob[order(res_allprob$higher),]$method
res_allprob_melt$method <- factor(res_allprob_melt$method,levels=rank1)

class_prob_bp <- ggplot(res_allprob_melt,aes_string(x="method",y="value",fill="variable"))+geom_bar(stat="identity")+ggtitle("Random Forest True Prediction Probability")+ylab("% of probability group")+coord_cartesian(ylim=c(0,1))+guides(fill=guide_legend(title="Probability"))+scale_fill_manual(values=c("darkorange","goldenrod1","dodgerblue"))+xlab("Method")+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))
class_prob_bp
```


# clustering
## summary and visualization
functions
```{r}
getCluster <- function(x,meta,npc=10,cent=4){
    tmp <- CreateSeuratObject(counts=x,meta.data=meta,min.cells=0,min.features=0)
    tmp <- FindVariableFeatures(tmp,nfeatures=3000,verbose=FALSE)
    tmp <- ScaleData(tmp,verbose=FALSE)
    tmp <- RunPCA(tmp,verbose=FALSE)
    tmp$kmeans <- kmeans(tmp@reductions$pca@cell.embeddings[,1:npc],centers=cent)$cluster
    tmpgraph <- buildSNNGraph(tmp@reductions$pca@cell.embeddings[,1:npc],transposed=T,k=10,d=NA)
    res <- cluster_louvain(tmpgraph)$membership
    cc <- aggregate(tmp@reductions$pca@cell.embeddings[,1:npc],list(res),mean)
    cc <- as.matrix(cc[,-1])
    hclu <- hclust(dist(cc))
    clu <- cutree(hclu,cent)
    clu <- clu[res]      
    tmp$lv_clu <- clu
    #tmp <- tmp@meta.data
    tmp <- RunUMAP(tmp,dims=1:npc,verbose=FALSE)
    return(tmp)
}

evalClustering <- function(x,npc=10,method="lv_clu"){
    acc <- -mean(sapply(unique(x@meta.data[,method]),function(i){
        p <- table(x@meta.data[x@meta.data[,method]==i,"celltype"])/sum(x@meta.data[,method]==i)
        sum(p*log(p))
    }))

    pur <- -mean(sapply(unique(x$celltype),function(sct){
        p <- table(x@meta.data[x$celltype==sct,method])/sum(x$celltype==sct)
        sum(p*log(p))
    }))

    ARI <- adjustedRandIndex(x$celltype,x@meta.data[,method])
    
    NMI_value <- NMI(x@meta.data[,method],x$celltype)
    #NMI_value <- NMI(x$lv_clu,x$celltype,variant="sum")
    
    df <- c(acc,pur,ARI,NMI_value)
    names(df) <- c("acc","pur","ARI","NMI")
    return(df)
}
```

process data
```{r}
dat1 <- lapply(dat,getCluster,metadat,cent=7)
dat2 <- lapply(dat2,getCluster,metadat2,cent=5)

res1_lv <- as.data.frame(lapply(dat1,evalClustering,method="lv_clu"))
#res1_lv <- as.data.frame(t(res1_lv))
res2_lv <- as.data.frame(lapply(dat2,evalClustering,method="lv_clu"))
#res2_lv <- as.data.frame(t(res2_lv))

res1_km <- as.data.frame(lapply(dat1,evalClustering,method="kmeans"))
#res1_km <- as.data.frame(t(res1_km))
res2_km <- as.data.frame(lapply(dat2,evalClustering,method="kmeans"))
#res2_km <- as.data.frame(t(res2_km))

res1_lv <- res1_lv[,colnames(res2_lv)]
res1_km <- res1_km[,colnames(res2_km)]
```


### process data lv
```{r}
##LV cluster
rownames(res1_lv) <- c("Hacc","Hpur","ARI","NMI")

res_lv <- data.frame()
res_lv_median <- data.frame()
for(i in 1:2){
    tmp1 <- res1_lv[i,1]-res1_lv[i,-1]
    tmp2 <- res2_lv[i,1]-res2_lv[i,-1]
    tmp <- rbind(tmp1,tmp2)
    rownames(tmp) <- paste0(rownames(res1_lv)[i],"_",c("GSE75748","GSE81861"))
    tmp_median <- apply(tmp,2,median)
    res_lv <- rbind(res_lv,tmp)
    res_lv_median <- rbind(res_lv_median,tmp_median)
}

for(i in 3:4){
    tmp1 <- res1_lv[i,-1]-res1_lv[i,1]
    tmp2 <- res2_lv[i,-1]-res2_lv[i,1]
    tmp <- rbind(tmp1,tmp2)
    rownames(tmp) <- paste0(rownames(res1_lv)[i],"_",c("GSE75748","GSE81861"))
    tmp_median <- apply(tmp,2,median)
    res_lv <- rbind(res_lv,tmp)
    res_lv_median <- rbind(res_lv_median,tmp_median)
}

res_lv_median <- as.data.frame(t(res_lv_median))
rownames(res_lv_median) <- colnames(res1_lv)[-1]
colnames(res_lv_median) <- c("Hacc","Hpur","ARI","NMI")

res_lv_median$method <- rownames(res_lv_median)
res_lv_median_melt <- data.table::melt(res_lv_median,id.vars=c("method"))
colnames(res_lv_median_melt)[2:3] <- c("Metrics","Metrics_Difference")
head(res_lv_median_melt)

res_lv$Metrics_Datasets <- rownames(res_lv)
res_lv$Metrics <- sapply(rownames(res_lv),function(x) strsplit(x,"\\_")[[1]][1] )
res_lv_melt <- data.table::melt(res_lv,id.vars=c("Metrics_Datasets","Metrics"))
colnames(res_lv_melt)[3:4] <- c("method","Metrics_Difference")
head(res_lv_melt)
```

heatmap(median metrics)
```{r}
rank1 <- apply(res_lv_median[,-ncol(res_lv_median)],2,rank)
rank1 <- apply(rank1,1,median)
rank1 <- names(rank1[order(rank1)])
#res_lv_median_melt$Method <- factor(res_lv_median_melt$method,levels=unique(res_lv_median_melt$method[order(res_lv_median_melt$method)]))
res_lv_median_melt$method <- factor(res_lv_median_melt$method,levels=rank1)
tmp_hm <- res_lv_median_melt[,]
tmp_hm[tmp_hm$Metrics_Difference<(-0.1),]$Metrics_Difference <- -0.1
#tmp_hm[tmp_hm$Metrics_Difference>(0.03),]$Metrics_Difference <- 0.03

clu_4metric_lv_hm <- ggplot(tmp_hm,aes(method,Metrics,fill=Metrics_Difference))+geom_tile()+scale_fill_gradient2(low="blue",mid="white",high="red",midpoint=0)+ggtitle("Louvain Clustering Evaluation Metrics(Values of Difference)")+xlab("Method")+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))
clu_4metric_lv_hm
```

### process data kmeans
```{r}
##LV cluster
rownames(res1_km) <- c("Hacc","Hpur","ARI","NMI")

res_km <- data.frame()
res_km_median <- data.frame()
for(i in 1:2){
    tmp1 <- res1_km[i,1]-res1_km[i,-1]
    tmp2 <- res2_km[i,1]-res2_km[i,-1]
    tmp <- rbind(tmp1,tmp2)
    rownames(tmp) <- paste0(rownames(res1_km)[i],"_",c("GSE75748","GSE81861"))
    tmp_median <- apply(tmp,2,median)
    res_km <- rbind(res_km,tmp)
    res_km_median <- rbind(res_km_median,tmp_median)
}

for(i in 3:4){
    tmp1 <- res1_km[i,-1]-res1_km[i,1]
    tmp2 <- res2_km[i,-1]-res2_km[i,1]
    tmp <- rbind(tmp1,tmp2)
    rownames(tmp) <- paste0(rownames(res1_km)[i],"_",c("GSE75748","GSE81861"))
    tmp_median <- apply(tmp,2,median)
    res_km <- rbind(res_km,tmp)
    res_km_median <- rbind(res_km_median,tmp_median)
}

res_km_median <- as.data.frame(t(res_km_median))
rownames(res_km_median) <- colnames(res1_km)[-1]
colnames(res_km_median) <- c("Hacc","Hpur","ARI","NMI")

res_km_median$method <- rownames(res_km_median)
res_km_median_melt <- data.table::melt(res_km_median,id.vars=c("method"))
colnames(res_km_median_melt)[2:3] <- c("Metrics","Metrics_Difference")
head(res_km_median_melt)

res_km$Metrics_Datasets <- rownames(res_km)
res_km$Metrics <- sapply(rownames(res_km),function(x) strsplit(x,"\\_")[[1]][1] )
res_km_melt <- data.table::melt(res_km,id.vars=c("Metrics_Datasets","Metrics"))
colnames(res_km_melt)[3:4] <- c("method","Metrics_Difference")
head(res_km_melt)
```

heatmap(median metrics)
```{r}
rank1 <- apply(res_km_median[,-ncol(res_km_median)],2,rank)
rank1 <- apply(rank1,1,median)
rank1 <- names(rank1[order(rank1)])
res_km_median_melt$method <- factor(res_km_median_melt$method,levels=rank1)

tmp_hm <- res_km_median_melt[,]
tmp_hm[tmp_hm$Metrics_Difference<(-0.1),]$Metrics_Difference <- -0.1
#tmp_hm[tmp_hm$Metrics_Difference>(0.01),]$Metrics_Difference <- 0.01

clu_4metric_km_hm <- ggplot(tmp_hm,aes(method,Metrics,fill=Metrics_Difference))+geom_tile()+scale_fill_gradient2(low="blue",mid="white",high="red",midpoint=0)+ggtitle("K-means Clustering Evaluation Metrics(Values of Difference)")+xlab("Method")+theme(plot.title=element_text(size=16),legend.title=element_text(size=15),legend.text=element_text(size=14),axis.title=element_text(size=18),axis.text.x=element_text(angle=90,size=15),axis.text.y=element_text(size=15))
clu_4metric_km_hm
```

# combine plot
```{r fig.height=9, fig.width=14}
ggarrange(de_p_all_scc_bp,de_fc_all_scc_bp,de_top_improve_bp_rst,de_fp_bp,class_acc_bp,class_prob_bp,pred_all_auc_bp,clu_4metric_lv_hm,clu_4metric_km_hm,scbulk_smd_scc_bp,pbbulk_smd_scc_bp,pbbulk_fc_scc_bp,ncol=3,nrow=4,labels="AUTO")
```



